{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "04705e5b",
   "metadata": {},
   "source": [
    "### Tutorial 4: Analysis of Starmap hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05213e8e",
   "metadata": {},
   "source": [
    "In this tutorial, we will conduct a downstream analysis of the reconstructed data obtained after training the stCAMBL model. The trained data can be obtained from the network disk. Please note that this is an R tutorial file, please run Celina analysis in the R environment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "370d3003",
   "metadata": {},
   "source": [
    "**Read the data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "03a00e6c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'count'</li><li>'meta'</li><li>'location'</li><li>'X_emb'</li><li>'X_recon'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'count'\n",
       "\\item 'meta'\n",
       "\\item 'location'\n",
       "\\item 'X\\_emb'\n",
       "\\item 'X\\_recon'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'count'\n",
       "2. 'meta'\n",
       "3. 'location'\n",
       "4. 'X_emb'\n",
       "5. 'X_recon'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"count\"    \"meta\"     \"location\" \"X_emb\"    \"X_recon\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# read raw data\n",
    "load('/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/hippocampus_scRNA_reference.RData')\n",
    "load('/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/starmap_plus_data_use.RData')\n",
    "# data trained by stCAMBL\n",
    "load('/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/Hippocampus_Analysis/hip_stCAMBL.RData')\n",
    "names(data_use_stCAMBL)\n",
    "expr_use <- data_use_stCAMBL$X_recon\n",
    "location_use <- data_use_stCAMBL$location\n",
    "celltype_use <- data_use$celltype_proportion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0169eb5",
   "metadata": {},
   "source": [
    "**Perform deconvolution**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d70f9dd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 methods overwritten by 'registry':\n",
      "  method               from \n",
      "  print.registry_field proxy\n",
      "  print.registry_entry proxy\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## QC on scRNASeq dataset! ...\n",
      "## QC on spatially-resolved dataset! ...\n",
      "## create reference matrix from scRNASeq...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "载入需要的程序包：nnls\n",
      "\n",
      "载入需要的程序包：ggplot2\n",
      "\n",
      "载入需要的程序包：TOAST\n",
      "\n",
      "载入需要的程序包：EpiDISH\n",
      "\n",
      "载入需要的程序包：limma\n",
      "\n",
      "\n",
      "载入程序包：‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "载入需要的程序包：quadprog\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "MuSiC v1.0.0 support SingleCellExperiment! See Tutorial: https://xuranw.github.io/MuSiC/articles/MuSiC.html\n",
      "\n",
      "MuSiC2 for multi-condition bulk RNA-seq data is also available!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Select Informative Genes! ...\n",
      "## Deconvolution Starts! ...\n",
      "## Deconvolution Finish! ...\n"
     ]
    }
   ],
   "source": [
    "options(warn = -1)\n",
    "library(CARD)\n",
    "location_use2 <- location_use \n",
    "colnames(location_use2) <- c(\"x\", \"y\")\n",
    "CARD_obj <- createCARDObject(\n",
    "  sc_count = scRNA_count_subset,\n",
    "  sc_meta = sc_meta_in_subset,\n",
    "  spatial_count = as.matrix(expr_use),\n",
    "  spatial_location = location_use2,\n",
    "  ct.varname = \"cellType\",\n",
    "  ct.select = unique(sc_meta_in_subset$cellType),\n",
    "  sample.varname = \"sampleInfo\",\n",
    "  minCountGene = 100,\n",
    "  minCountSpot = 20\n",
    ")\n",
    "\n",
    "CARD_obj <- CARD_deconvolution(CARD_object = CARD_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab3739b",
   "metadata": {},
   "source": [
    "**Perform CELINA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1de91696",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtering genes based on threshold = 5e-05 \n",
      "Get gene list for each cell type.\n",
      " [1] \"Interneuron\"                \"Entorhinal cortex\"         \n",
      " [3] \"Deep layer subiculum\"       \"Entorhinal cortex (IEG)\"   \n",
      " [5] \"Subiculum\"                  \"Dentate Principal cells\"   \n",
      " [7] \"Medial entorrhinal cortex\"  \"Medial entorrhinal cortexm\"\n",
      " [9] \"Postsubiculum\"              \"CA3 Principal cells\"       \n",
      "[11] \"Astrocyte\"                  \"Oligodendrocyte\"           \n",
      "[13] \"MyelinProcesses\"            \"Microglia\"                 \n",
      "[15] \"Resident macrophage\"        \"Ependymal\"                 \n",
      "[17] \"Endothelial_Stalk\"         \n"
     ]
    }
   ],
   "source": [
    "library(CELINA)\n",
    "Obj <- Create_Celina_Object(\n",
    "  celltype_mat = t(celltype_use),             # spot × celltype\n",
    "  gene_expression_mat = as.matrix(expr_use),             # gene × spot\n",
    "  location = as.matrix(location_use),     # spot × 2, class must be data.frame!\n",
    "  covariates = NULL,\n",
    "  project = \"starmap\"\n",
    ")\n",
    "# if you want to further filter cell types based on their total proportion across spots, or you only want to test a subset of cell types, you can select the cell types here:\n",
    "filtered_cell_types = colnames(data_use$celltype_proportion)[which(colSums(data_use$celltype_proportion) >150)]\n",
    "\n",
    "Obj = preprocess_input(Obj, \n",
    "                       # Celina object\n",
    "                       cell_types_to_test = filtered_cell_types,  \n",
    "                       # a vector of cell types to be used for testing \n",
    "                       scRNA_count = as.matrix(scRNA_count_subset), \n",
    "                       # a gene x cell expression matrix of reference scRNA-seq data\n",
    "                       sc_cell_type_labels = sc_meta_in_subset$cellType,\n",
    "                       # a vector of cell type labels for each cell in scRNA_count\n",
    "                       threshold = 5e-5)\n",
    "print(names(Obj@genes_list))\n",
    "Obj <- Calculate_Kernel(Obj, approximation = FALSE)\n",
    "\n",
    "Obj = Testing_interaction_all(Obj,celltype_to_test = 'Microglia', num_cores=10)\n",
    "save(Obj, file = \"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/Hippocampus_Analysis/CELINA_Obj_workspace.RData\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c25220fa",
   "metadata": {},
   "source": [
    "**Screen genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "91c2512e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Microglia'"
      ],
      "text/latex": [
       "'Microglia'"
      ],
      "text/markdown": [
       "'Microglia'"
      ],
      "text/plain": [
       "[1] \"Microglia\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(CELINA)\n",
    "tmp_env <- new.env()\n",
    "load(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/Hippocampus_Analysis/CELINA_Obj_workspace.RData\", envir = tmp_env)\n",
    "  \n",
    "Obj <- tmp_env$Obj\n",
    "names(Obj@result)\n",
    "for (celltype in names(Obj@result)) {\n",
    "  colnames(Obj@result[[celltype]]) <- \"pval\"\n",
    "}\n",
    "\n",
    "Significant_gene_list_Celina <- list()\n",
    "fdr_threshold <- 0.00005\n",
    "for (celltype in names(Obj@result)) {\n",
    "  df <- Obj@result[[celltype]]\n",
    "  if (is.null(df) || nrow(df) == 0) next\n",
    "  pvals <- df[[1]]\n",
    "  fdr <- p.adjust(pvals, method = \"BH\")\n",
    "  sig_genes <- rownames(df)[which(fdr < fdr_threshold)]\n",
    "  Significant_gene_list_Celina[[celltype]] <- sig_genes\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1250689e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Abl1'</li><li>'Actr2'</li><li>'Adam10'</li><li>'Anpep'</li><li>'Arhgap39'</li><li>'Arhgdia'</li><li>'Arhgef2'</li><li>'Arrb2'</li><li>'Atp6v1f'</li><li>'Baz1a'</li><li>'Bin1'</li><li>'C1qa'</li><li>'Ccl6'</li><li>'Ccnd1'</li><li>'Ccr2'</li><li>'Cd81'</li><li>'Cd83'</li><li>'Cd9'</li><li>'Cdc42'</li><li>'Cfl1'</li><li>'Clcf1'</li><li>'Creb1'</li><li>'Csrnp1'</li><li>'Cst3'</li><li>'Cst7'</li><li>'Ctsl'</li><li>'Ctss'</li><li>'Cul1'</li><li>'Cxcl10'</li><li>'Dbnl'</li><li>'Dnm2'</li><li>'Dock10'</li><li>'Dtnbp1'</li><li>'Ebf3'</li><li>'Eef2k'</li><li>'Elavl1'</li><li>'Fcgr1'</li><li>'Fcrls'</li><li>'Ftl1'</li><li>'Fubp1'</li><li>'Gdf15'</li><li>'Gdi2'</li><li>'Golm1'</li><li>'Grn'</li><li>'Hexb'</li><li>'Hnrnpa3'</li><li>'Hnrnpab'</li><li>'Hnrnph1'</li><li>'Hnrnpk'</li><li>'Hsd17b4'</li><li>'Itgb5'</li><li>'Itm2b'</li><li>'Itpr2'</li><li>'Kansl1'</li><li>'Kcnj2'</li><li>'Kctd12'</li><li>'Lhfpl2'</li><li>'Lpcat2'</li><li>'Lrpap1'</li><li>'Lrpprc'</li><li>'Ly86'</li><li>'Lyn'</li><li>'Maml3'</li><li>'Manf'</li><li>'Mapk14'</li><li>'Marcks'</li><li>'Mertk'</li><li>'Nav3'</li><li>'Nf1'</li><li>'Nr3c1'</li><li>'Nufip1'</li><li>'Numb'</li><li>'Olfml3'</li><li>'Ophn1'</li><li>'P2ry12'</li><li>'Pabpc1'</li><li>'Pacsin2'</li><li>'Pak2'</li><li>'Picalm'</li><li>'Pld1'</li><li>'Plekho1'</li><li>'Ppp1ca'</li><li>'Ppp1cc'</li><li>'Ptpra'</li><li>'Ptpro'</li><li>'Pum2'</li><li>'Rab14'</li><li>'Rab1b'</li><li>'Rab5a'</li><li>'Rac1'</li><li>'Rap1a'</li><li>'Rela'</li><li>'Rgs2'</li><li>'Rin3'</li><li>'Rinl'</li><li>'Sdcbp'</li><li>'Sdhb'</li><li>'Selplg'</li><li>'Sema4d'</li><li>'Serpine2'</li><li>'Sfpq'</li><li>'Smad7'</li><li>'Sparc'</li><li>'Srrm2'</li><li>'Stag1'</li><li>'Stat3'</li><li>'Stau1'</li><li>'Stk40'</li><li>'Tapbp'</li><li>'Tbxas1'</li><li>'Tfap2c'</li><li>'Timp2'</li><li>'Tmem176b'</li><li>'Tmsb4x'</li><li>'Trem2'</li><li>'Ttr'</li><li>'Ubc'</li><li>'Vapa'</li><li>'Vcp'</li><li>'Vps16'</li><li>'Vps33b'</li><li>'Whrn'</li><li>'Xbp1'</li><li>'Zc3h7a'</li><li>'Zcwpw1'</li><li>'Zdhhc5'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Abl1'\n",
       "\\item 'Actr2'\n",
       "\\item 'Adam10'\n",
       "\\item 'Anpep'\n",
       "\\item 'Arhgap39'\n",
       "\\item 'Arhgdia'\n",
       "\\item 'Arhgef2'\n",
       "\\item 'Arrb2'\n",
       "\\item 'Atp6v1f'\n",
       "\\item 'Baz1a'\n",
       "\\item 'Bin1'\n",
       "\\item 'C1qa'\n",
       "\\item 'Ccl6'\n",
       "\\item 'Ccnd1'\n",
       "\\item 'Ccr2'\n",
       "\\item 'Cd81'\n",
       "\\item 'Cd83'\n",
       "\\item 'Cd9'\n",
       "\\item 'Cdc42'\n",
       "\\item 'Cfl1'\n",
       "\\item 'Clcf1'\n",
       "\\item 'Creb1'\n",
       "\\item 'Csrnp1'\n",
       "\\item 'Cst3'\n",
       "\\item 'Cst7'\n",
       "\\item 'Ctsl'\n",
       "\\item 'Ctss'\n",
       "\\item 'Cul1'\n",
       "\\item 'Cxcl10'\n",
       "\\item 'Dbnl'\n",
       "\\item 'Dnm2'\n",
       "\\item 'Dock10'\n",
       "\\item 'Dtnbp1'\n",
       "\\item 'Ebf3'\n",
       "\\item 'Eef2k'\n",
       "\\item 'Elavl1'\n",
       "\\item 'Fcgr1'\n",
       "\\item 'Fcrls'\n",
       "\\item 'Ftl1'\n",
       "\\item 'Fubp1'\n",
       "\\item 'Gdf15'\n",
       "\\item 'Gdi2'\n",
       "\\item 'Golm1'\n",
       "\\item 'Grn'\n",
       "\\item 'Hexb'\n",
       "\\item 'Hnrnpa3'\n",
       "\\item 'Hnrnpab'\n",
       "\\item 'Hnrnph1'\n",
       "\\item 'Hnrnpk'\n",
       "\\item 'Hsd17b4'\n",
       "\\item 'Itgb5'\n",
       "\\item 'Itm2b'\n",
       "\\item 'Itpr2'\n",
       "\\item 'Kansl1'\n",
       "\\item 'Kcnj2'\n",
       "\\item 'Kctd12'\n",
       "\\item 'Lhfpl2'\n",
       "\\item 'Lpcat2'\n",
       "\\item 'Lrpap1'\n",
       "\\item 'Lrpprc'\n",
       "\\item 'Ly86'\n",
       "\\item 'Lyn'\n",
       "\\item 'Maml3'\n",
       "\\item 'Manf'\n",
       "\\item 'Mapk14'\n",
       "\\item 'Marcks'\n",
       "\\item 'Mertk'\n",
       "\\item 'Nav3'\n",
       "\\item 'Nf1'\n",
       "\\item 'Nr3c1'\n",
       "\\item 'Nufip1'\n",
       "\\item 'Numb'\n",
       "\\item 'Olfml3'\n",
       "\\item 'Ophn1'\n",
       "\\item 'P2ry12'\n",
       "\\item 'Pabpc1'\n",
       "\\item 'Pacsin2'\n",
       "\\item 'Pak2'\n",
       "\\item 'Picalm'\n",
       "\\item 'Pld1'\n",
       "\\item 'Plekho1'\n",
       "\\item 'Ppp1ca'\n",
       "\\item 'Ppp1cc'\n",
       "\\item 'Ptpra'\n",
       "\\item 'Ptpro'\n",
       "\\item 'Pum2'\n",
       "\\item 'Rab14'\n",
       "\\item 'Rab1b'\n",
       "\\item 'Rab5a'\n",
       "\\item 'Rac1'\n",
       "\\item 'Rap1a'\n",
       "\\item 'Rela'\n",
       "\\item 'Rgs2'\n",
       "\\item 'Rin3'\n",
       "\\item 'Rinl'\n",
       "\\item 'Sdcbp'\n",
       "\\item 'Sdhb'\n",
       "\\item 'Selplg'\n",
       "\\item 'Sema4d'\n",
       "\\item 'Serpine2'\n",
       "\\item 'Sfpq'\n",
       "\\item 'Smad7'\n",
       "\\item 'Sparc'\n",
       "\\item 'Srrm2'\n",
       "\\item 'Stag1'\n",
       "\\item 'Stat3'\n",
       "\\item 'Stau1'\n",
       "\\item 'Stk40'\n",
       "\\item 'Tapbp'\n",
       "\\item 'Tbxas1'\n",
       "\\item 'Tfap2c'\n",
       "\\item 'Timp2'\n",
       "\\item 'Tmem176b'\n",
       "\\item 'Tmsb4x'\n",
       "\\item 'Trem2'\n",
       "\\item 'Ttr'\n",
       "\\item 'Ubc'\n",
       "\\item 'Vapa'\n",
       "\\item 'Vcp'\n",
       "\\item 'Vps16'\n",
       "\\item 'Vps33b'\n",
       "\\item 'Whrn'\n",
       "\\item 'Xbp1'\n",
       "\\item 'Zc3h7a'\n",
       "\\item 'Zcwpw1'\n",
       "\\item 'Zdhhc5'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Abl1'\n",
       "2. 'Actr2'\n",
       "3. 'Adam10'\n",
       "4. 'Anpep'\n",
       "5. 'Arhgap39'\n",
       "6. 'Arhgdia'\n",
       "7. 'Arhgef2'\n",
       "8. 'Arrb2'\n",
       "9. 'Atp6v1f'\n",
       "10. 'Baz1a'\n",
       "11. 'Bin1'\n",
       "12. 'C1qa'\n",
       "13. 'Ccl6'\n",
       "14. 'Ccnd1'\n",
       "15. 'Ccr2'\n",
       "16. 'Cd81'\n",
       "17. 'Cd83'\n",
       "18. 'Cd9'\n",
       "19. 'Cdc42'\n",
       "20. 'Cfl1'\n",
       "21. 'Clcf1'\n",
       "22. 'Creb1'\n",
       "23. 'Csrnp1'\n",
       "24. 'Cst3'\n",
       "25. 'Cst7'\n",
       "26. 'Ctsl'\n",
       "27. 'Ctss'\n",
       "28. 'Cul1'\n",
       "29. 'Cxcl10'\n",
       "30. 'Dbnl'\n",
       "31. 'Dnm2'\n",
       "32. 'Dock10'\n",
       "33. 'Dtnbp1'\n",
       "34. 'Ebf3'\n",
       "35. 'Eef2k'\n",
       "36. 'Elavl1'\n",
       "37. 'Fcgr1'\n",
       "38. 'Fcrls'\n",
       "39. 'Ftl1'\n",
       "40. 'Fubp1'\n",
       "41. 'Gdf15'\n",
       "42. 'Gdi2'\n",
       "43. 'Golm1'\n",
       "44. 'Grn'\n",
       "45. 'Hexb'\n",
       "46. 'Hnrnpa3'\n",
       "47. 'Hnrnpab'\n",
       "48. 'Hnrnph1'\n",
       "49. 'Hnrnpk'\n",
       "50. 'Hsd17b4'\n",
       "51. 'Itgb5'\n",
       "52. 'Itm2b'\n",
       "53. 'Itpr2'\n",
       "54. 'Kansl1'\n",
       "55. 'Kcnj2'\n",
       "56. 'Kctd12'\n",
       "57. 'Lhfpl2'\n",
       "58. 'Lpcat2'\n",
       "59. 'Lrpap1'\n",
       "60. 'Lrpprc'\n",
       "61. 'Ly86'\n",
       "62. 'Lyn'\n",
       "63. 'Maml3'\n",
       "64. 'Manf'\n",
       "65. 'Mapk14'\n",
       "66. 'Marcks'\n",
       "67. 'Mertk'\n",
       "68. 'Nav3'\n",
       "69. 'Nf1'\n",
       "70. 'Nr3c1'\n",
       "71. 'Nufip1'\n",
       "72. 'Numb'\n",
       "73. 'Olfml3'\n",
       "74. 'Ophn1'\n",
       "75. 'P2ry12'\n",
       "76. 'Pabpc1'\n",
       "77. 'Pacsin2'\n",
       "78. 'Pak2'\n",
       "79. 'Picalm'\n",
       "80. 'Pld1'\n",
       "81. 'Plekho1'\n",
       "82. 'Ppp1ca'\n",
       "83. 'Ppp1cc'\n",
       "84. 'Ptpra'\n",
       "85. 'Ptpro'\n",
       "86. 'Pum2'\n",
       "87. 'Rab14'\n",
       "88. 'Rab1b'\n",
       "89. 'Rab5a'\n",
       "90. 'Rac1'\n",
       "91. 'Rap1a'\n",
       "92. 'Rela'\n",
       "93. 'Rgs2'\n",
       "94. 'Rin3'\n",
       "95. 'Rinl'\n",
       "96. 'Sdcbp'\n",
       "97. 'Sdhb'\n",
       "98. 'Selplg'\n",
       "99. 'Sema4d'\n",
       "100. 'Serpine2'\n",
       "101. 'Sfpq'\n",
       "102. 'Smad7'\n",
       "103. 'Sparc'\n",
       "104. 'Srrm2'\n",
       "105. 'Stag1'\n",
       "106. 'Stat3'\n",
       "107. 'Stau1'\n",
       "108. 'Stk40'\n",
       "109. 'Tapbp'\n",
       "110. 'Tbxas1'\n",
       "111. 'Tfap2c'\n",
       "112. 'Timp2'\n",
       "113. 'Tmem176b'\n",
       "114. 'Tmsb4x'\n",
       "115. 'Trem2'\n",
       "116. 'Ttr'\n",
       "117. 'Ubc'\n",
       "118. 'Vapa'\n",
       "119. 'Vcp'\n",
       "120. 'Vps16'\n",
       "121. 'Vps33b'\n",
       "122. 'Whrn'\n",
       "123. 'Xbp1'\n",
       "124. 'Zc3h7a'\n",
       "125. 'Zcwpw1'\n",
       "126. 'Zdhhc5'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"Abl1\"     \"Actr2\"    \"Adam10\"   \"Anpep\"    \"Arhgap39\" \"Arhgdia\" \n",
       "  [7] \"Arhgef2\"  \"Arrb2\"    \"Atp6v1f\"  \"Baz1a\"    \"Bin1\"     \"C1qa\"    \n",
       " [13] \"Ccl6\"     \"Ccnd1\"    \"Ccr2\"     \"Cd81\"     \"Cd83\"     \"Cd9\"     \n",
       " [19] \"Cdc42\"    \"Cfl1\"     \"Clcf1\"    \"Creb1\"    \"Csrnp1\"   \"Cst3\"    \n",
       " [25] \"Cst7\"     \"Ctsl\"     \"Ctss\"     \"Cul1\"     \"Cxcl10\"   \"Dbnl\"    \n",
       " [31] \"Dnm2\"     \"Dock10\"   \"Dtnbp1\"   \"Ebf3\"     \"Eef2k\"    \"Elavl1\"  \n",
       " [37] \"Fcgr1\"    \"Fcrls\"    \"Ftl1\"     \"Fubp1\"    \"Gdf15\"    \"Gdi2\"    \n",
       " [43] \"Golm1\"    \"Grn\"      \"Hexb\"     \"Hnrnpa3\"  \"Hnrnpab\"  \"Hnrnph1\" \n",
       " [49] \"Hnrnpk\"   \"Hsd17b4\"  \"Itgb5\"    \"Itm2b\"    \"Itpr2\"    \"Kansl1\"  \n",
       " [55] \"Kcnj2\"    \"Kctd12\"   \"Lhfpl2\"   \"Lpcat2\"   \"Lrpap1\"   \"Lrpprc\"  \n",
       " [61] \"Ly86\"     \"Lyn\"      \"Maml3\"    \"Manf\"     \"Mapk14\"   \"Marcks\"  \n",
       " [67] \"Mertk\"    \"Nav3\"     \"Nf1\"      \"Nr3c1\"    \"Nufip1\"   \"Numb\"    \n",
       " [73] \"Olfml3\"   \"Ophn1\"    \"P2ry12\"   \"Pabpc1\"   \"Pacsin2\"  \"Pak2\"    \n",
       " [79] \"Picalm\"   \"Pld1\"     \"Plekho1\"  \"Ppp1ca\"   \"Ppp1cc\"   \"Ptpra\"   \n",
       " [85] \"Ptpro\"    \"Pum2\"     \"Rab14\"    \"Rab1b\"    \"Rab5a\"    \"Rac1\"    \n",
       " [91] \"Rap1a\"    \"Rela\"     \"Rgs2\"     \"Rin3\"     \"Rinl\"     \"Sdcbp\"   \n",
       " [97] \"Sdhb\"     \"Selplg\"   \"Sema4d\"   \"Serpine2\" \"Sfpq\"     \"Smad7\"   \n",
       "[103] \"Sparc\"    \"Srrm2\"    \"Stag1\"    \"Stat3\"    \"Stau1\"    \"Stk40\"   \n",
       "[109] \"Tapbp\"    \"Tbxas1\"   \"Tfap2c\"   \"Timp2\"    \"Tmem176b\" \"Tmsb4x\"  \n",
       "[115] \"Trem2\"    \"Ttr\"      \"Ubc\"      \"Vapa\"     \"Vcp\"      \"Vps16\"   \n",
       "[121] \"Vps33b\"   \"Whrn\"     \"Xbp1\"     \"Zc3h7a\"   \"Zcwpw1\"   \"Zdhhc5\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Significant_gene_list_Celina[['Microglia']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a0324ae",
   "metadata": {},
   "source": [
    "**Visualize cell type distribution, gene expression, and Aβ plaques**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "660dc5a0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "载入需要的程序包：SingleCellExperiment\n",
      "\n",
      "载入需要的程序包：SummarizedExperiment\n",
      "\n",
      "载入需要的程序包：MatrixGenerics\n",
      "\n",
      "载入需要的程序包：matrixStats\n",
      "\n",
      "\n",
      "载入程序包：‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:wrMisc’:\n",
      "\n",
      "    colSds, rowSds\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "载入程序包：‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:wrMisc’:\n",
      "\n",
      "    colSds, rowSds\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "载入需要的程序包：GenomicRanges\n",
      "\n",
      "载入需要的程序包：stats4\n",
      "\n",
      "载入需要的程序包：S4Vectors\n",
      "\n",
      "\n",
      "载入程序包：‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "载入需要的程序包：IRanges\n",
      "\n",
      "载入需要的程序包：GenomeInfoDb\n",
      "\n",
      "载入需要的程序包：scuttle\n",
      "\n",
      "\n",
      "载入程序包：‘scater’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMDS\n",
      "\n",
      "\n",
      "载入需要的程序包：viridisLite\n",
      "\n",
      "Plotting cell type:Microglia\n",
      "\n",
      "Plotting gene:P2ry12\n",
      "\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mcolour\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mcolour\u001b[39m, which will replace the existing scale.\n",
      "Plot saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_P2ry12_comprehensive.pdf\n",
      "\n",
      "Average distance from cells to nearest plaque:327.26\n",
      "\n",
      "Median distance from cells to nearest plaque:261.5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(scater)\n",
    "library(ggpubr)\n",
    "library(pdist)\n",
    "library(viridis)\n",
    "\n",
    "# Data preprocessing\n",
    "expr_use_scater_norm <- normalizeCounts(expr_use)\n",
    "plaque <- read.csv(\"/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/plaque_13months-disease-replicate_1.csv\")\n",
    "\n",
    "# Calculate distance from each cell to the nearest plaque\n",
    "calculate_plaque_distance <- function(location_data, plaque_data) {\n",
    "  dist_p <- c()\n",
    "  for(cc in 1:nrow(location_data)){\n",
    "    dist_p[cc] <- min(pdist(location_data[cc,], plaque_data[,1:2])@dist)\n",
    "  }\n",
    "  return(dist_p)\n",
    "}\n",
    "\n",
    "# Plotting function\n",
    "plot_factor_value <- function(location, feature, textmethod, pointsize = 2, textsize = 15,\n",
    "                             shape = 15, lowcolor = \"#05a8aa\", midcolor = \"#edede9\",\n",
    "                             highcolor = \"#bc412b\", featurename = \"Feature\") {\n",
    "  location <- as.data.frame(location)\n",
    "  datt <- data.frame(feature = feature, x = location[, 1], y = location[, 2])\n",
    "  \n",
    "  p <- ggplot(datt, aes(x = x, y = y, color = feature)) +\n",
    "    geom_point(size = pointsize, alpha = 0.9, shape = shape) +\n",
    "    scale_color_gradient2(low = lowcolor, mid = midcolor, high = highcolor) +\n",
    "    ggtitle(textmethod) +\n",
    "    theme_void() +\n",
    "    theme(plot.title = element_text(size = textsize),\n",
    "          text = element_text(size = textsize),\n",
    "          legend.position = \"right\",\n",
    "          legend.title = element_text(family = \"serif\"),\n",
    "          legend.text = element_text(family = \"serif\")) +\n",
    "    labs(color = featurename)\n",
    "  \n",
    "  return(p)\n",
    "}\n",
    "\n",
    "# Plaque plotting function\n",
    "plot_plaque_cells <- function(location_data, plaque_data, textsize = 15) {\n",
    "  # Prepare plaque data\n",
    "  datt_plaque <- plaque_data[, c(\"m.cx\", \"m.cy\", \"s.radius.mean\")]\n",
    "  colnames(datt_plaque) <- c(\"x\", \"y\", \"radius\")\n",
    "  datt_plaque$Type <- \"Plaque\"\n",
    "  datt_plaque$Size <- (plaque_data$s.radius.mean - min(plaque_data$s.radius.mean)) / \n",
    "                     (max(plaque_data$s.radius.mean) - min(plaque_data$s.radius.mean)) * 4\n",
    "  \n",
    "  # Prepare cell data\n",
    "  dattt_cells <- location_data\n",
    "  dattt_cells$radius <- 0.5\n",
    "  dattt_cells$Size <- 0.5\n",
    "  colnames(dattt_cells) <- c(\"x\", \"y\", \"radius\", \"Size\")\n",
    "  dattt_cells$Type <- \"Cell\"\n",
    "  \n",
    "  # Combine data\n",
    "  dat_fig <- rbind(dattt_cells, datt_plaque)\n",
    "  dat_fig$Type <- factor(dat_fig$Type, levels = c(\"Plaque\", \"Cell\"), ordered = TRUE)\n",
    "  \n",
    "  # Plot\n",
    "  p <- ggplot() +\n",
    "    scale_color_viridis(discrete = TRUE) +\n",
    "    geom_point(data = dat_fig, aes(x = x, y = y, color = Type, shape = Type), \n",
    "               alpha = 0.9, size = dat_fig$Size) + \n",
    "    ggtitle(\"Plaques & Cells Distribution\") + \n",
    "    theme_void() + \n",
    "    theme(plot.title = element_text(size = textsize, family = 'serif'), \n",
    "          text = element_text(size = textsize, family = 'serif'), \n",
    "          legend.position = \"right\",\n",
    "          legend.title = element_text(family = \"serif\"),\n",
    "          legend.text = element_text(family = \"serif\")) + \n",
    "    labs(color = NULL, shape = NULL) +\n",
    "    guides(colour = guide_legend(override.aes = list(size = 4)))\n",
    "  \n",
    "  return(p)\n",
    "}\n",
    "\n",
    "# Comprehensive plotting function\n",
    "plot_comprehensive_spatial <- function(celltype, gene_name, \n",
    "                                     plaque_file_path = NULL,\n",
    "                                     output_file = \"Data/hippocampus/comprehensive_spatial.pdf\",\n",
    "                                     width = 18, height = 6) {\n",
    "  \n",
    "  # Check if cell type exists\n",
    "  if (!celltype %in% colnames(data_use$celltype_proportion)) {\n",
    "    stop(paste(\"Cell type\", celltype, \"does not exist in the data\"))\n",
    "  }\n",
    "  \n",
    "  # Check if gene exists\n",
    "  if (!gene_name %in% rownames(expr_use_scater_norm)) {\n",
    "    stop(paste(\"Gene\", gene_name, \"does not exist in the expression matrix\"))\n",
    "  }\n",
    "  \n",
    "  message(\"Plotting cell type:\", celltype)\n",
    "  message(\"Plotting gene:\", gene_name)\n",
    "  \n",
    "  # Plot cell type proportion\n",
    "  prop <- data_use$celltype_proportion[, celltype]\n",
    "  prop_size <- ((prop - min(prop)) / (max(prop) - min(prop))) * 3\n",
    "  p1 <- plot_factor_value(location_use, prop,\n",
    "                          textmethod = paste(celltype, \"Proportion\"),\n",
    "                          pointsize = prop_size,\n",
    "                          textsize = 14,\n",
    "                          shape = 16,\n",
    "                          lowcolor = \"#05a8aa\",\n",
    "                          midcolor = \"#edede9\",\n",
    "                          highcolor = \"#bc412b\",\n",
    "                          featurename = \"Proportion\") +\n",
    "    scale_color_viridis(option = \"C\") +\n",
    "    labs(color = NULL)\n",
    "  \n",
    "  # Plot gene expression\n",
    "  expr <- expr_use_scater_norm[gene_name, ]\n",
    "  expr_size <- ((expr - min(expr)) / (max(expr) - min(expr))) * 3\n",
    "  p2 <- plot_factor_value(location_use, expr,\n",
    "                          textmethod = paste(gene_name, \"Expression\"),\n",
    "                          pointsize = expr_size,\n",
    "                          textsize = 14,\n",
    "                          shape = 16,\n",
    "                          lowcolor = \"#faedcd\",\n",
    "                          midcolor = \"#edede9\",\n",
    "                          highcolor = \"#0077b6\",\n",
    "                          featurename = \"Expression\") +\n",
    "    scale_color_viridis(option = \"D\") +\n",
    "    labs(color = NULL)\n",
    "  \n",
    "  # Plot plaque distribution\n",
    "  if (!is.null(plaque_file_path)) {\n",
    "    plaque_data <- read.csv(plaque_file_path)\n",
    "  } else {\n",
    "    # Use global variable plaque\n",
    "    if (exists(\"plaque\")) {\n",
    "      plaque_data <- plaque\n",
    "    } else {\n",
    "      stop(\"Please provide plaque data file path or ensure plaque data is loaded\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  p3 <- plot_plaque_cells(data_use$location_use, plaque_data, textsize = 14)\n",
    "  \n",
    "  # Combine plots\n",
    "  combined_plot <- ggarrange(p1, p2, p3, ncol = 3, nrow = 1)\n",
    "  \n",
    "  # Save PDF\n",
    "  ggsave(output_file, combined_plot, \n",
    "         width = width, height = height, \n",
    "         device = \"pdf\")\n",
    "  \n",
    "  message(\"Plot saved to:\", output_file)\n",
    "  \n",
    "  # Calculate and display distance information\n",
    "  dist_p <- calculate_plaque_distance(data_use$location_use, plaque_data[,1:2])\n",
    "  message(\"Average distance from cells to nearest plaque:\", round(mean(dist_p), 2))\n",
    "  message(\"Median distance from cells to nearest plaque:\", round(median(dist_p), 2))\n",
    "  \n",
    "  return(list(plot = combined_plot, distances = dist_p))\n",
    "}\n",
    "\n",
    "# Simplified usage function\n",
    "quick_comprehensive_plot <- function(celltype, gene, filename = NULL, plaque_path = NULL) {\n",
    "  \n",
    "  # Auto-generate filename if not specified\n",
    "  if (is.null(filename)) {\n",
    "    filename <- paste0(\"comprehensive_\", celltype, \"_\", gene, \".pdf\")\n",
    "  }\n",
    "  \n",
    "  result <- plot_comprehensive_spatial(\n",
    "    celltype = celltype,\n",
    "    gene_name = gene,\n",
    "    plaque_file_path = plaque_path,\n",
    "    output_file = filename,\n",
    "    width = 18,\n",
    "    height = 6\n",
    "  )\n",
    "  \n",
    "  return(result)\n",
    "}\n",
    "\n",
    "# Call function\n",
    "# Basic usage\n",
    "result <- plot_comprehensive_spatial(\n",
    "  celltype = \"Microglia\",           \n",
    "  gene_name = \"P2ry12\",            \n",
    "  plaque_file_path = \"/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/plaque_13months-disease-replicate_1.csv\",\n",
    "  output_file = \"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_P2ry12_comprehensive.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cd39d2c",
   "metadata": {},
   "source": [
    "**Plot the distribution of stCAMBL reconstruction data at a distance from the plaque for specific cell types**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82d000d0",
   "metadata": {},
   "source": [
    "Microglia often cluster around amyloid-β plaques, including so-called disease-associated microglia (DAMs), which are highly expressive of genes associated with phagocytosis and inflammation, such as Grn. You can also replace gene names to analyze other genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1238882",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Analyzing stCAMBL data - Cell type specific analysis\n",
      "\n",
      "Analyzing gene:Trem2\n",
      "\n",
      "Cell type:Microglia\n",
      "\n",
      "Distance groups:0-15, 20-40, 50-80, >500\n",
      "\n",
      "Cell type threshold: 0.1\n",
      "\n",
      "Number of qualified cells: 1768\n",
      "\n",
      "Total number of cells: 9244\n",
      "\n",
      "Cell type proportion range: 0 - 0.933\n",
      "\n",
      "Target cell type group statistics:\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 5 × 5\u001b[39m\n",
      "  distance_group count mean_expr median_expr mean_celltype_prop\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m              \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m 20-40um          131     0.688       0.672              0.475\n",
      "\u001b[90m2\u001b[39m 50-80um          177     0.678       0.653              0.393\n",
      "\u001b[90m3\u001b[39m >500um           251     0.556       0.537              0.343\n",
      "\u001b[90m4\u001b[39m Other           \u001b[4m1\u001b[24m195     0.591       0.575              0.303\n",
      "\u001b[90m5\u001b[39m ≤15um             14     0.661       0.675              0.553\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Combined plot saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Trem2_Microglia_detailed.pdf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(scater)\n",
    "library(ggpubr)\n",
    "library(pdist)\n",
    "library(viridis)\n",
    "library(dplyr)\n",
    "library(RColorBrewer)\n",
    "\n",
    "PLAQUE_COLOR <- \"#8B4B9F\" \n",
    "plaque = read.csv(\"/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/plaque_13months-disease-replicate_1.csv\")\n",
    "\n",
    "# Data preprocessing\n",
    "expr_use_scater_norm <- normalizeCounts(expr_use)\n",
    "\n",
    "# Cell distance classification function\n",
    "classify_cells_by_plaque_distance_celltype <- function(location_data, plaque_data, \n",
    "                                                     celltype_proportion, celltype_name,\n",
    "                                                     celltype_threshold = 0.5,  # Cell type threshold\n",
    "                                                     distance_groups = list(c(10, 20), c(50, Inf)),\n",
    "                                                     group_names = NULL) {\n",
    "  # Check if cell type exists\n",
    "  if (!celltype_name %in% colnames(celltype_proportion)) {\n",
    "    stop(paste(\"Cell type\", celltype_name, \"does not exist in cell type proportion matrix\"))\n",
    "  }\n",
    "  \n",
    "  cell_info <- data.frame(\n",
    "    x = location_data[,1],\n",
    "    y = location_data[,2],\n",
    "    cell_id = 1:nrow(location_data),\n",
    "    min_distance = NA,\n",
    "    nearest_plaque_id = NA,\n",
    "    distance_group = \"Other\",  \n",
    "    celltype_prop = celltype_proportion[, celltype_name],  # Cell type proportion\n",
    "    is_target_celltype = FALSE  \n",
    "  )\n",
    "  \n",
    "  # Filter target cell types\n",
    "  cell_info$is_target_celltype <- cell_info$celltype_prop >= celltype_threshold\n",
    "  message(paste(\"Cell type threshold:\", celltype_threshold))\n",
    "  message(paste(\"Number of qualified cells:\", sum(cell_info$is_target_celltype)))\n",
    "  message(paste(\"Total number of cells:\", nrow(cell_info)))\n",
    "  message(paste(\"Cell type proportion range:\", round(min(cell_info$celltype_prop), 3), \"-\", round(max(cell_info$celltype_prop), 3)))\n",
    "  \n",
    "  # Calculate distance from each cell to all plaques\n",
    "  for(i in 1:nrow(cell_info)) {\n",
    "    distances <- pdist(location_data[i,], plaque_data[,1:2])@dist\n",
    "    cell_info$min_distance[i] <- min(distances)\n",
    "    cell_info$nearest_plaque_id[i] <- which.min(distances)\n",
    "  }\n",
    "  \n",
    "  # Distance grouping for target cell types\n",
    "  target_cells <- which(cell_info$is_target_celltype)\n",
    "  \n",
    "  # Group by distance intervals\n",
    "  for(i in 1:length(distance_groups)) {\n",
    "    group_range <- distance_groups[[i]]\n",
    "    lower_bound <- group_range[1]\n",
    "    upper_bound <- group_range[2]\n",
    "    \n",
    "    # Create group names\n",
    "    if(!is.null(group_names) && length(group_names) >= i) {\n",
    "      group_name <- group_names[i]\n",
    "    } else {\n",
    "      if(is.infinite(upper_bound)) {\n",
    "        group_name <- paste0(\">\", lower_bound, \"um\")\n",
    "      } else if(lower_bound == 0) {\n",
    "        group_name <- paste0(\"≤\", upper_bound, \"um\")\n",
    "      } else {\n",
    "        group_name <- paste0(lower_bound, \"-\", upper_bound, \"um\")\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    # Group target cell types\n",
    "    if(is.infinite(upper_bound)) {\n",
    "      in_group <- target_cells[cell_info$min_distance[target_cells] > lower_bound]\n",
    "    } else {\n",
    "      in_group <- target_cells[cell_info$min_distance[target_cells] >= lower_bound & \n",
    "                              cell_info$min_distance[target_cells] <= upper_bound]\n",
    "    }\n",
    "    \n",
    "    cell_info$distance_group[in_group] <- group_name\n",
    "  }\n",
    "  \n",
    "  # Add numeric group labels for subsequent analysis\n",
    "  unique_groups <- unique(cell_info$distance_group[cell_info$is_target_celltype])\n",
    "  cell_info$group_numeric <- NA\n",
    "  cell_info$group_numeric[cell_info$is_target_celltype] <- as.numeric(factor(\n",
    "    cell_info$distance_group[cell_info$is_target_celltype], \n",
    "    levels = unique_groups\n",
    "  ))\n",
    "  \n",
    "  return(cell_info)\n",
    "}\n",
    "\n",
    "# Create plaque boundary circle data\n",
    "create_plaque_circles <- function(plaque_data, n_points = 100) {\n",
    "  circle_data <- data.frame()\n",
    "  \n",
    "  for(i in 1:nrow(plaque_data)) {\n",
    "    # Create circle points\n",
    "    theta <- seq(0, 2*pi, length.out = n_points)\n",
    "    x_circle <- plaque_data$m.cx[i] + plaque_data$s.radius.mean[i] * cos(theta)\n",
    "    y_circle <- plaque_data$m.cy[i] + plaque_data$s.radius.mean[i] * sin(theta)\n",
    "    \n",
    "    temp_circle <- data.frame(\n",
    "      x = x_circle,\n",
    "      y = y_circle,\n",
    "      plaque_id = i,\n",
    "      radius = plaque_data$s.radius.mean[i]\n",
    "    )\n",
    "    \n",
    "    circle_data <- rbind(circle_data, temp_circle)\n",
    "  }\n",
    "  \n",
    "  return(circle_data)\n",
    "}\n",
    "\n",
    "# Modified main plotting function - adding cell type analysis\n",
    "plot_plaque_celltype_distance_expression <- function(gene_name, \n",
    "                                                    celltype_name,  # Cell type name\n",
    "                                                    celltype_threshold = 0.5,  # Cell type threshold\n",
    "                                                    plaque_file_path = NULL,\n",
    "                                                    distance_groups = list(c(0, 20), c(50, Inf)),\n",
    "                                                    group_names = NULL,\n",
    "                                                    output_file = NULL,\n",
    "                                                    save_individual = TRUE,\n",
    "                                                    show_non_target_cells = TRUE, \n",
    "                                                    width = 14, height = 10) {\n",
    "  \n",
    "  # Check if gene exists\n",
    "  if (!gene_name %in% rownames(expr_use_scater_norm)) {\n",
    "    stop(paste(\"Gene\", gene_name, \"does not exist in expression matrix\"))\n",
    "  }\n",
    "  \n",
    "  # Check if cell type exists\n",
    "  if (!celltype_name %in% colnames(data_use$celltype_proportion)) {\n",
    "    available_types <- colnames(data_use$celltype_proportion)\n",
    "    stop(paste(\"Cell type\", celltype_name, \"does not exist in cell type proportion matrix.\\nAvailable cell types:\", \n",
    "               paste(available_types, collapse = \", \")))\n",
    "  }\n",
    "  \n",
    "  # Read plaque data\n",
    "  if (!is.null(plaque_file_path)) {\n",
    "    plaque_data <- read.csv(plaque_file_path)\n",
    "  } else {\n",
    "    if (exists(\"plaque\")) {\n",
    "      plaque_data <- plaque\n",
    "    } else {\n",
    "      stop(\"Please provide plaque data file path or ensure plaque data is loaded\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  message(\"Analyzing stCAMBL data - Cell type specific analysis\")\n",
    "  message(\"Analyzing gene:\", gene_name)\n",
    "  message(\"Cell type:\", celltype_name)\n",
    "  message(\"Distance groups:\", paste(sapply(distance_groups, function(x) {\n",
    "    if(is.infinite(x[2])) paste0(\">\", x[1]) else paste0(x[1], \"-\", x[2])\n",
    "  }), collapse = \", \"))\n",
    "  \n",
    "  # Classify cells\n",
    "  cell_info <- classify_cells_by_plaque_distance_celltype(\n",
    "    data_use$location_use, \n",
    "    plaque_data, \n",
    "    data_use$celltype_proportion,\n",
    "    celltype_name,\n",
    "    celltype_threshold,\n",
    "    distance_groups, \n",
    "    group_names\n",
    "  )\n",
    "  \n",
    "  # Get gene expression data\n",
    "  gene_expr <- expr_use_scater_norm[gene_name, ]\n",
    "  cell_info$gene_expression <- gene_expr\n",
    "  \n",
    "  # Create plaque boundary circles\n",
    "  plaque_circles <- create_plaque_circles(plaque_data)\n",
    "  \n",
    "  # Prepare plotting data\n",
    "  if (show_non_target_cells) {\n",
    "    plot_data <- cell_info\n",
    "    # Set different transparency and size for different cell types\n",
    "    plot_data$alpha_val <- ifelse(!plot_data$is_target_celltype, 0.1, \n",
    "                                 ifelse(plot_data$distance_group == \"Other\", 0.3, 0.8))\n",
    "    plot_data$size_val <- ifelse(!plot_data$is_target_celltype, 0.3, \n",
    "                                ifelse(plot_data$distance_group == \"Other\", 0.8, 1.5))\n",
    "  } else {\n",
    "    # Only show target cell types\n",
    "    plot_data <- cell_info[cell_info$is_target_celltype, ]\n",
    "    plot_data$alpha_val <- ifelse(plot_data$distance_group == \"Other\", 0.5, 0.9)\n",
    "    plot_data$size_val <- ifelse(plot_data$distance_group == \"Other\", 1.0, 1.8)\n",
    "  }\n",
    "  \n",
    "  # Get all groups in target cell type and set colors\n",
    "  target_cell_data <- cell_info[cell_info$is_target_celltype, ]\n",
    "  all_groups <- unique(target_cell_data$distance_group)\n",
    "  n_groups <- length(all_groups)\n",
    "  \n",
    "  # Set colors for different groups\n",
    "  if(n_groups <= 8) {\n",
    "    group_colors <- brewer.pal(max(3, n_groups), \"Set2\")[1:n_groups]\n",
    "  } else {\n",
    "    group_colors <- rainbow(n_groups)\n",
    "  }\n",
    "  names(group_colors) <- all_groups\n",
    "  \n",
    "  # Add gray for non-target cells if showing them\n",
    "  if (show_non_target_cells) {\n",
    "    group_colors[\"Non_Target\"] <- \"lightgray\"\n",
    "    plot_data$display_group <- ifelse(plot_data$is_target_celltype, \n",
    "                                     plot_data$distance_group, \"Non_Target\")\n",
    "  } else {\n",
    "    plot_data$display_group <- plot_data$distance_group\n",
    "  }\n",
    "  \n",
    "  # Statistics\n",
    "  group_stats <- target_cell_data %>%\n",
    "    group_by(distance_group) %>%\n",
    "    summarise(\n",
    "      count = n(),\n",
    "      mean_expr = mean(gene_expression, na.rm = TRUE),\n",
    "      median_expr = median(gene_expression, na.rm = TRUE),\n",
    "      mean_celltype_prop = mean(celltype_prop, na.rm = TRUE),\n",
    "      .groups = 'drop'\n",
    "    )\n",
    "  \n",
    "  message(\"Target cell type group statistics:\")\n",
    "  print(group_stats)\n",
    "  \n",
    "  # Create main plot 1: Cell type proportion plot\n",
    "  p1 <- ggplot() +\n",
    "    # Plot cell points colored by cell type proportion\n",
    "    geom_point(data = plot_data, \n",
    "               aes(x = x, y = y, color = celltype_prop, size = display_group, alpha = display_group)) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color and size settings\n",
    "    scale_color_viridis_c(option = \"A\", name = paste(celltype_name, \"\\nProportion\")) +\n",
    "    scale_size_manual(values = setNames(plot_data$size_val[match(unique(plot_data$display_group), plot_data$display_group)], \n",
    "                                       unique(plot_data$display_group)), guide = \"none\") +\n",
    "    scale_alpha_manual(values = setNames(plot_data$alpha_val[match(unique(plot_data$display_group), plot_data$display_group)], \n",
    "                                        unique(plot_data$display_group)), guide = \"none\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(1.0, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(celltype_name, \" Distribution\")) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create main plot 2: Distance group plot\n",
    "  target_plot_data <- plot_data[plot_data$is_target_celltype, ]\n",
    "  \n",
    "  p2 <- ggplot() +\n",
    "    # Plot cell points colored by distance group\n",
    "    geom_point(data = target_plot_data, \n",
    "               aes(x = x, y = y, color = distance_group), \n",
    "               size = 1.5, alpha = 0.8) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color settings\n",
    "    scale_color_manual(values = group_colors, name = \"Distance Group\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(1.0, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(celltype_name, \" Distance Groups\")) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create spatial gene expression plot\n",
    "  p3 <- ggplot() +\n",
    "    # Plot cell points colored by gene expression\n",
    "    geom_point(data = target_plot_data, \n",
    "               aes(x = x, y = y, color = gene_expression, shape = distance_group), \n",
    "               size = 1.8, alpha = 0.8) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color settings\n",
    "    scale_color_viridis_c(option = \"D\", name = paste(gene_name, \"\\nExpression\")) +\n",
    "    scale_shape_manual(values = c(16, 17, 15, 18, 19, 8, 11, 12)[1:n_groups], name = \"Distance Group\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(0.8, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(gene_name, \" in \", celltype_name)) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create expression comparison boxplot\n",
    "  plot_data_for_box <- target_cell_data[target_cell_data$distance_group != \"Other\", ]\n",
    "  \n",
    "  p4 <- ggplot(plot_data_for_box, aes(x = distance_group, y = gene_expression, fill = distance_group)) +\n",
    "    geom_boxplot(alpha = 0.7) +\n",
    "    geom_jitter(width = 0.2, alpha = 0.6, size = 1.0) +\n",
    "    scale_fill_manual(values = group_colors) +\n",
    "    labs(x = \"Distance Group\", y = paste(gene_name, \"Expression\"), \n",
    "         title = paste(gene_name, \"in\", celltype_name)) +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      text = element_text(size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      axis.title.x = element_text(size = 26),\n",
    "      axis.title.y = element_text(size = 26),\n",
    "      axis.text.x = element_text(size = 26, angle = 0, hjust = 0.5),\n",
    "      axis.text.y = element_text(size = 26),\n",
    "      legend.position = \"none\"\n",
    "    ) +\n",
    "    stat_compare_means(method = \"kruskal.test\", label = \"p.format\", \n",
    "                       size = 5)\n",
    "  \n",
    "  # Combine plots\n",
    "  combined_plot <- ggarrange(\n",
    "    ggarrange(p1, p2, ncol = 2, widths = c(1, 1)),\n",
    "    ggarrange(p3, p4, ncol = 2, widths = c(1, 1)),\n",
    "    nrow = 2, heights = c(1, 1)\n",
    "  )\n",
    "  \n",
    "  # Save plots\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, combined_plot, \n",
    "           width = width, height = height, \n",
    "           device = \"pdf\")\n",
    "    message(\"Combined plot saved to:\", output_file)\n",
    "    \n",
    "    if (save_individual) {\n",
    "      file_parts <- tools::file_path_sans_ext(output_file)\n",
    "      file_ext <- tools::file_ext(output_file)\n",
    "      if (file_ext == \"\") file_ext <- \"pdf\"\n",
    "      \n",
    "      # Save individual subplots\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_celltype_distribution.\", \"_no_csl\", file_ext), p1, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_distance_groups.\",\"_no_csl\", file_ext), p2, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_expression_spatial.\",\"_no_csl\", file_ext), p3, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_expression_comparison.\",\"_no_csl\",  file_ext), p4, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Return results\n",
    "  return(list(\n",
    "    combined_plot = combined_plot,\n",
    "    celltype_distribution_plot = p1,\n",
    "    distance_groups_plot = p2,\n",
    "    expression_spatial_plot = p3,\n",
    "    expression_comparison_plot = p4,\n",
    "    cell_info = cell_info,\n",
    "    target_cell_stats = group_stats,\n",
    "    celltype_summary = list(\n",
    "      total_cells = nrow(cell_info),\n",
    "      target_cells = sum(cell_info$is_target_celltype),\n",
    "      target_percentage = round(sum(cell_info$is_target_celltype) / nrow(cell_info) * 100, 2)\n",
    "    )\n",
    "  ))\n",
    "}\n",
    "\n",
    "\n",
    "# Set cell type name\n",
    "cc <- \"Microglia\"  \n",
    "result_celltype_detailed <- plot_plaque_celltype_distance_expression(\n",
    "  # Picalm, Grn, Cfl1 ,Trem2\n",
    "  gene_name = \"Grn\",\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.1,\n",
    "  distance_groups = list(c(0, 15), c(20, 40), c(50, 80), c(500, Inf)),\n",
    "\n",
    "  output_file = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Grn_\", cc, \"_detailed.pdf\"),\n",
    "  save_individual = FALSE,\n",
    "  show_non_target_cells = FALSE,  \n",
    "  width = 16,\n",
    "  height = 12\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a631b135",
   "metadata": {},
   "source": [
    "For comparison, we do the same analysis of the raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55b08ef4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Analyzing Raw data - Cell type specific analysis\n",
      "\n",
      "Analyzing gene:Trem2\n",
      "\n",
      "Cell type:Microglia\n",
      "\n",
      "Distance groups:0-15, 20-40, 50-80, >100\n",
      "\n",
      "Cell type threshold: 0.1\n",
      "\n",
      "Number of qualified cells: 1768\n",
      "\n",
      "Total number of cells: 9244\n",
      "\n",
      "Cell type proportion range: 0 - 0.933\n",
      "\n",
      "Target cell type group statistics:\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 5 × 5\u001b[39m\n",
      "  distance_group count mean_expr median_expr mean_celltype_prop\n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m              \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m\n",
      "\u001b[90m1\u001b[39m 20-40um          131     1.21         1.28              0.475\n",
      "\u001b[90m2\u001b[39m 50-80um          177     0.801        0                 0.393\n",
      "\u001b[90m3\u001b[39m >100um          \u001b[4m1\u001b[24m292     0.241        0                 0.299\n",
      "\u001b[90m4\u001b[39m Other            154     0.678        0                 0.398\n",
      "\u001b[90m5\u001b[39m ≤15um             14     1.64         1.61              0.553\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Combined plot saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Trem2_Microglia_detailed_raw.pdf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(scater)\n",
    "library(ggpubr)\n",
    "library(pdist)\n",
    "library(viridis)\n",
    "library(dplyr)\n",
    "library(RColorBrewer)\n",
    "\n",
    "PLAQUE_COLOR <- \"#8B4B9F\"  \n",
    "plaque = read.csv(\"/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/plaque_13months-disease-replicate_1.csv\")\n",
    "\n",
    "# Data preprocessing\n",
    "expr_use_scater_norm_raw <- normalizeCounts(data_use$count_use)\n",
    "\n",
    "# Cell distance classification function\n",
    "classify_cells_by_plaque_distance_celltype <- function(location_data, plaque_data, \n",
    "                                                     celltype_proportion, celltype_name,\n",
    "                                                     celltype_threshold = 0.5,  \n",
    "                                                     distance_groups = list(c(10, 20), c(50, Inf)),\n",
    "                                                     group_names = NULL) {\n",
    "  # Check if cell type exists\n",
    "  if (!celltype_name %in% colnames(celltype_proportion)) {\n",
    "    stop(paste(\"Cell type\", celltype_name, \"does not exist in cell type proportion matrix\"))\n",
    "  }\n",
    "  \n",
    "  cell_info <- data.frame(\n",
    "    x = location_data[,1],\n",
    "    y = location_data[,2],\n",
    "    cell_id = 1:nrow(location_data),\n",
    "    min_distance = NA,\n",
    "    nearest_plaque_id = NA,\n",
    "    distance_group = \"Other\", \n",
    "    celltype_prop = celltype_proportion[, celltype_name], \n",
    "    is_target_celltype = FALSE  # Whether it is target cell type\n",
    "  )\n",
    "  \n",
    "  # Filter target cell types\n",
    "  cell_info$is_target_celltype <- cell_info$celltype_prop >= celltype_threshold\n",
    "  \n",
    "  message(paste(\"Cell type threshold:\", celltype_threshold))\n",
    "  message(paste(\"Number of qualified cells:\", sum(cell_info$is_target_celltype)))\n",
    "  message(paste(\"Total number of cells:\", nrow(cell_info)))\n",
    "  message(paste(\"Cell type proportion range:\", round(min(cell_info$celltype_prop), 3), \"-\", round(max(cell_info$celltype_prop), 3)))\n",
    "  \n",
    "  # Calculate distance from each cell to all plaques\n",
    "  for(i in 1:nrow(cell_info)) {\n",
    "    distances <- pdist(location_data[i,], plaque_data[,1:2])@dist\n",
    "    cell_info$min_distance[i] <- min(distances)\n",
    "    cell_info$nearest_plaque_id[i] <- which.min(distances)\n",
    "  }\n",
    "  \n",
    "  # Distance grouping for target cell types\n",
    "  target_cells <- which(cell_info$is_target_celltype)\n",
    "  \n",
    "  # Group by distance intervals\n",
    "  for(i in 1:length(distance_groups)) {\n",
    "    group_range <- distance_groups[[i]]\n",
    "    lower_bound <- group_range[1]\n",
    "    upper_bound <- group_range[2]\n",
    "    \n",
    "    # Create group names\n",
    "    if(!is.null(group_names) && length(group_names) >= i) {\n",
    "      group_name <- group_names[i]\n",
    "    } else {\n",
    "      if(is.infinite(upper_bound)) {\n",
    "        group_name <- paste0(\">\", lower_bound, \"um\")\n",
    "      } else if(lower_bound == 0) {\n",
    "        group_name <- paste0(\"≤\", upper_bound, \"um\")\n",
    "      } else {\n",
    "        group_name <- paste0(lower_bound, \"-\", upper_bound, \"um\")\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    # Group target cell types\n",
    "    if(is.infinite(upper_bound)) {\n",
    "      in_group <- target_cells[cell_info$min_distance[target_cells] > lower_bound]\n",
    "    } else {\n",
    "      in_group <- target_cells[cell_info$min_distance[target_cells] >= lower_bound & \n",
    "                              cell_info$min_distance[target_cells] <= upper_bound]\n",
    "    }\n",
    "    \n",
    "    cell_info$distance_group[in_group] <- group_name\n",
    "  }\n",
    "  \n",
    "  # Add numeric group labels for subsequent analysis\n",
    "  unique_groups <- unique(cell_info$distance_group[cell_info$is_target_celltype])\n",
    "  cell_info$group_numeric <- NA\n",
    "  cell_info$group_numeric[cell_info$is_target_celltype] <- as.numeric(factor(\n",
    "    cell_info$distance_group[cell_info$is_target_celltype], \n",
    "    levels = unique_groups\n",
    "  ))\n",
    "  \n",
    "  return(cell_info)\n",
    "}\n",
    "\n",
    "# Create plaque boundary circle data\n",
    "create_plaque_circles <- function(plaque_data, n_points = 100) {\n",
    "  circle_data <- data.frame()\n",
    "  \n",
    "  for(i in 1:nrow(plaque_data)) {\n",
    "    # Create circle points\n",
    "    theta <- seq(0, 2*pi, length.out = n_points)\n",
    "    x_circle <- plaque_data$m.cx[i] + plaque_data$s.radius.mean[i] * cos(theta)\n",
    "    y_circle <- plaque_data$m.cy[i] + plaque_data$s.radius.mean[i] * sin(theta)\n",
    "    \n",
    "    temp_circle <- data.frame(\n",
    "      x = x_circle,\n",
    "      y = y_circle,\n",
    "      plaque_id = i,\n",
    "      radius = plaque_data$s.radius.mean[i]\n",
    "    )\n",
    "    \n",
    "    circle_data <- rbind(circle_data, temp_circle)\n",
    "  }\n",
    "  \n",
    "  return(circle_data)\n",
    "}\n",
    "\n",
    "# Plotting function\n",
    "plot_plaque_celltype_distance_expression <- function(gene_name, \n",
    "                                                    celltype_name,  # Cell type name\n",
    "                                                    celltype_threshold = 0.5,  # Cell type threshold\n",
    "                                                    plaque_file_path = NULL,\n",
    "                                                    distance_groups = list(c(0, 20), c(50, Inf)),\n",
    "                                                    group_names = NULL,\n",
    "                                                    output_file = NULL,\n",
    "                                                    save_individual = TRUE,\n",
    "                                                    show_non_target_cells = TRUE,  # Whether to show non-target cell types\n",
    "                                                    width = 14, height = 10) {\n",
    "  \n",
    "  # Check if gene exists\n",
    "  if (!gene_name %in% rownames(expr_use_scater_norm_raw)) {\n",
    "    stop(paste(\"Gene\", gene_name, \"does not exist in expression matrix\"))\n",
    "  }\n",
    "  \n",
    "  # Check if cell type exists\n",
    "  if (!celltype_name %in% colnames(data_use$celltype_proportion)) {\n",
    "    available_types <- colnames(data_use$celltype_proportion)\n",
    "    stop(paste(\"Cell type\", celltype_name, \"does not exist in cell type proportion matrix.\\nAvailable cell types:\", \n",
    "               paste(available_types, collapse = \", \")))\n",
    "  }\n",
    "  \n",
    "  # Read plaque data\n",
    "  if (!is.null(plaque_file_path)) {\n",
    "    plaque_data <- read.csv(plaque_file_path)\n",
    "  } else {\n",
    "    if (exists(\"plaque\")) {\n",
    "      plaque_data <- plaque\n",
    "    } else {\n",
    "      stop(\"Please provide plaque data file path or ensure plaque data is loaded\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  message(\"Analyzing Raw data - Cell type specific analysis\")\n",
    "  message(\"Analyzing gene:\", gene_name)\n",
    "  message(\"Cell type:\", celltype_name)\n",
    "  message(\"Distance groups:\", paste(sapply(distance_groups, function(x) {\n",
    "    if(is.infinite(x[2])) paste0(\">\", x[1]) else paste0(x[1], \"-\", x[2])\n",
    "  }), collapse = \", \"))\n",
    "  \n",
    "  # Classify cells\n",
    "  cell_info <- classify_cells_by_plaque_distance_celltype(\n",
    "    data_use$location_use, \n",
    "    plaque_data, \n",
    "    data_use$celltype_proportion,\n",
    "    celltype_name,\n",
    "    celltype_threshold,\n",
    "    distance_groups, \n",
    "    group_names\n",
    "  )\n",
    "  \n",
    "  # Get gene expression data\n",
    "  gene_expr <- expr_use_scater_norm_raw[gene_name, ]\n",
    "  cell_info$gene_expression <- gene_expr\n",
    "  \n",
    "  # Create plaque boundary circles\n",
    "  plaque_circles <- create_plaque_circles(plaque_data)\n",
    "  \n",
    "  # Prepare plotting data\n",
    "  if (show_non_target_cells) {\n",
    "    plot_data <- cell_info\n",
    "    # Set different transparency and size for different cell types\n",
    "    plot_data$alpha_val <- ifelse(!plot_data$is_target_celltype, 0.1, \n",
    "                                 ifelse(plot_data$distance_group == \"Other\", 0.3, 0.8))\n",
    "    plot_data$size_val <- ifelse(!plot_data$is_target_celltype, 0.3, \n",
    "                                ifelse(plot_data$distance_group == \"Other\", 0.8, 1.5))\n",
    "  } else {\n",
    "    # Show target cell types\n",
    "    plot_data <- cell_info[cell_info$is_target_celltype, ]\n",
    "    plot_data$alpha_val <- ifelse(plot_data$distance_group == \"Other\", 0.5, 0.9)\n",
    "    plot_data$size_val <- ifelse(plot_data$distance_group == \"Other\", 1.0, 1.8)\n",
    "  }\n",
    "  \n",
    "  # Get all groups in target cell type and set colors\n",
    "  target_cell_data <- cell_info[cell_info$is_target_celltype, ]\n",
    "  all_groups <- unique(target_cell_data$distance_group)\n",
    "  n_groups <- length(all_groups)\n",
    "  \n",
    "  # Set colors for different groups\n",
    "  if(n_groups <= 8) {\n",
    "    group_colors <- brewer.pal(max(3, n_groups), \"Set2\")[1:n_groups]\n",
    "  } else {\n",
    "    group_colors <- rainbow(n_groups)\n",
    "  }\n",
    "  names(group_colors) <- all_groups\n",
    "  \n",
    "  # Add gray for non-target cells if showing them\n",
    "  if (show_non_target_cells) {\n",
    "    group_colors[\"Non_Target\"] <- \"lightgray\"\n",
    "    plot_data$display_group <- ifelse(plot_data$is_target_celltype, \n",
    "                                     plot_data$distance_group, \"Non_Target\")\n",
    "  } else {\n",
    "    plot_data$display_group <- plot_data$distance_group\n",
    "  }\n",
    "  \n",
    "  # Statistics\n",
    "  group_stats <- target_cell_data %>%\n",
    "    group_by(distance_group) %>%\n",
    "    summarise(\n",
    "      count = n(),\n",
    "      mean_expr = mean(gene_expression, na.rm = TRUE),\n",
    "      median_expr = median(gene_expression, na.rm = TRUE),\n",
    "      mean_celltype_prop = mean(celltype_prop, na.rm = TRUE),\n",
    "      .groups = 'drop'\n",
    "    )\n",
    "  \n",
    "  message(\"Target cell type group statistics:\")\n",
    "  print(group_stats)\n",
    "  \n",
    "  # Create main plot 1: Cell type proportion plot\n",
    "  p1 <- ggplot() +\n",
    "    # Plot cell points colored by cell type proportion\n",
    "    geom_point(data = plot_data, \n",
    "               aes(x = x, y = y, color = celltype_prop, size = display_group, alpha = display_group)) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color and size settings\n",
    "    scale_color_viridis_c(option = \"A\", name = paste(celltype_name, \"\\nProportion\")) +\n",
    "    scale_size_manual(values = setNames(plot_data$size_val[match(unique(plot_data$display_group), plot_data$display_group)], \n",
    "                                       unique(plot_data$display_group)), guide = \"none\") +\n",
    "    scale_alpha_manual(values = setNames(plot_data$alpha_val[match(unique(plot_data$display_group), plot_data$display_group)], \n",
    "                                        unique(plot_data$display_group)), guide = \"none\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(1.0, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(celltype_name, \" Distribution\")) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create main plot 2: Distance group plot\n",
    "  target_plot_data <- plot_data[plot_data$is_target_celltype, ]\n",
    "  \n",
    "  p2 <- ggplot() +\n",
    "    # Plot cell points colored by distance group\n",
    "    geom_point(data = target_plot_data, \n",
    "               aes(x = x, y = y, color = distance_group), \n",
    "               size = 1.5, alpha = 0.8) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color settings\n",
    "    scale_color_manual(values = group_colors, name = \"Distance Group\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(1.0, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(celltype_name, \" Distance Groups\")) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create spatial gene expression plot\n",
    "  p3 <- ggplot() +\n",
    "    # Plot cell points colored by gene expression\n",
    "    geom_point(data = target_plot_data, \n",
    "               aes(x = x, y = y, color = gene_expression, shape = distance_group), \n",
    "               size = 1.8, alpha = 0.8) +\n",
    "    # Plot plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = PLAQUE_COLOR, size = 1.2, alpha = 0.8) +\n",
    "    # Color settings\n",
    "    scale_color_viridis_c(option = \"D\", name = paste(gene_name, \"\\nExpression\")) +\n",
    "    scale_shape_manual(values = c(16, 17, 15, 18, 19, 8, 11, 12)[1:n_groups], name = \"Distance Group\") +\n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.title = element_text(size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.key.size = unit(0.8, \"cm\")\n",
    "    ) +\n",
    "    ggtitle(paste0(gene_name, \" in \", celltype_name)) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Create expression comparison boxplot\n",
    "  plot_data_for_box <- target_cell_data[target_cell_data$distance_group != \"Other\", ]\n",
    "  \n",
    "  p4 <- ggplot(plot_data_for_box, aes(x = distance_group, y = gene_expression, fill = distance_group)) +\n",
    "    geom_boxplot(alpha = 0.7) +\n",
    "    geom_jitter(width = 0.2, alpha = 0.6, size = 1.0) +\n",
    "    scale_fill_manual(values = group_colors) +\n",
    "    labs(x = \"Distance Group\", y = paste(gene_name, \"Expression\"), \n",
    "         title = paste(gene_name, \"in\", celltype_name)) +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      text = element_text(size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      axis.title.x = element_text(size = 26),\n",
    "      axis.title.y = element_text(size = 26),\n",
    "      axis.text.x = element_text(size = 26, angle = 0, hjust = 0.5),\n",
    "      axis.text.y = element_text(size = 26),\n",
    "      legend.position = \"none\"\n",
    "    ) +\n",
    "    stat_compare_means(method = \"kruskal.test\", label = \"p.format\", \n",
    "                       size = 5)\n",
    "  \n",
    "  # Combine plots\n",
    "  combined_plot <- ggarrange(\n",
    "    ggarrange(p1, p2, ncol = 2, widths = c(1, 1)),\n",
    "    ggarrange(p3, p4, ncol = 2, widths = c(1, 1)),\n",
    "    nrow = 2, heights = c(1, 1)\n",
    "  )\n",
    "  \n",
    "  # Save plots\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, combined_plot, \n",
    "           width = width, height = height, \n",
    "           device = \"pdf\")\n",
    "    message(\"Combined plot saved to:\", output_file)\n",
    "    \n",
    "    if (save_individual) {\n",
    "      file_parts <- tools::file_path_sans_ext(output_file)\n",
    "      file_ext <- tools::file_ext(output_file)\n",
    "      if (file_ext == \"\") file_ext <- \"pdf\"\n",
    "      \n",
    "      # Save individual subplots\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_celltype_distribution_raw.\", file_ext), p1, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_distance_groups_raw.\", file_ext), p2, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_expression_spatial_raw.\", file_ext), p3, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "      ggsave(paste0(file_parts, \"_\", gsub(\"[^A-Za-z0-9]\", \"_\", celltype_name), \"_expression_comparison_raw.\", file_ext), p4, \n",
    "             width = width * 0.5, height = height * 0.5, device = \"pdf\")\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Return results\n",
    "  return(list(\n",
    "    combined_plot = combined_plot,\n",
    "    celltype_distribution_plot = p1,\n",
    "    distance_groups_plot = p2,\n",
    "    expression_spatial_plot = p3,\n",
    "    expression_comparison_plot = p4,\n",
    "    cell_info = cell_info,\n",
    "    target_cell_stats = group_stats,\n",
    "    celltype_summary = list(\n",
    "      total_cells = nrow(cell_info),\n",
    "      target_cells = sum(cell_info$is_target_celltype),\n",
    "      target_percentage = round(sum(cell_info$is_target_celltype) / nrow(cell_info) * 100, 2)\n",
    "    )\n",
    "  ))\n",
    "}\n",
    "\n",
    "# Call function\n",
    "result_celltype_detailed <- plot_plaque_celltype_distance_expression(\n",
    "  # Picalm, Grn, Cfl1 ,Trem2\n",
    "  gene_name = \"Grn\",\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.1,\n",
    "  distance_groups = list(c(0, 15), c(20, 40), c(50, 80), c(100, Inf)),\n",
    "  output_file = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Grn_\", cc, \"_detailed_raw.pdf\"),\n",
    "  save_individual = FALSE,\n",
    "  show_non_target_cells = FALSE, \n",
    "  width = 16,\n",
    "  height = 12\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e111f3c",
   "metadata": {},
   "source": [
    "**Analysis of the relationship between gene expression and distance to Aβ plaques**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f40219c0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "corrplot 0.95 loaded\n",
      "\n",
      "Creating expression heatmap forGrninMicroglia...\n",
      "\n",
      "Using 748 Microglia cells (threshold: 0.3 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Heatmap saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Grn_Microglia_heatmap.pdf\n",
      "\n",
      "Analyzing gene expression and plaque distance correlation inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells\n",
      "\n",
      "=== Gene-Plaque Distance Correlation Results inMicroglia===\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    gene correlation abs_correlation\n",
      "1  Trem2  -0.2954035       0.2954035\n",
      "2    Grn  -0.2932961       0.2932961\n",
      "3 Picalm  -0.2794015       0.2794015\n",
      "4   Cfl1   0.1074841       0.1074841\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "Correlation analysis results saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_gene_distance_correlations.pdf\n",
      "\n",
      "Creating multi-gene heatmap comparison inMicroglia...\n",
      "\n",
      "Creating expression heatmap forPicalminMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forCfl1inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forGrninMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forTrem2inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Multi-gene heatmap saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_multi_gene_heatmaps.pdf\n",
      "\n",
      "=== Starting comprehensive plaque analysis forMicroglia===\n",
      "\n",
      "1. Generating expression heatmaps...\n",
      "\n",
      "Creating multi-gene heatmap comparison inMicroglia...\n",
      "\n",
      "Creating expression heatmap forPicalminMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forCfl1inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forGrninMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Creating expression heatmap forTrem2inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells (threshold: 0.1 )\n",
      "\n",
      "Calculating grid expression values...\n",
      "\n",
      "Multi-gene heatmap saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_comprehensive_analysis/analysis_Microglia/expression_heatmaps.pdf\n",
      "\n",
      "2. Performing gene-distance correlation analysis...\n",
      "\n",
      "Analyzing gene expression and plaque distance correlation inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells\n",
      "\n",
      "=== Gene-Plaque Distance Correlation Results inMicroglia===\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    gene correlation abs_correlation\n",
      "1  Trem2  -0.2954035       0.2954035\n",
      "2    Grn  -0.2932961       0.2932961\n",
      "3 Picalm  -0.2794015       0.2794015\n",
      "4   Cfl1   0.1074841       0.1074841\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "Correlation analysis results saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_comprehensive_analysis/analysis_Microglia/gene_distance_correlation.pdf\n",
      "\n",
      "3. Performing gene co-expression network analysis...\n",
      "\n",
      "Analyzing gene co-expression network inMicroglia...\n",
      "\n",
      "Using 1768 Microglia cells\n",
      "\n",
      "=== Strongly correlated gene pairs inMicroglia(|r| >= 0.3) ===\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Gene1 Gene2 Correlation Abs_Correlation\n",
      "1    Grn Trem2   0.9997741       0.9997741\n",
      "2 Picalm   Grn   0.9580008       0.9580008\n",
      "3 Picalm Trem2   0.9540979       0.9540979\n",
      "4 Picalm  Cfl1  -0.4255421       0.4255421\n",
      "5   Cfl1   Grn  -0.3559943       0.3559943\n",
      "6   Cfl1 Trem2  -0.3544326       0.3544326\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Gene network analysis results saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_comprehensive_analysis/analysis_Microglia/gene_network_analysis.pdf\n",
      "\n",
      "=== Analysis completed ===\n",
      "\n",
      "Results saved to:/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Microglia_comprehensive_analysis/analysis_Microglia\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"=== Analysis completed ===\"\n",
      "[1] \"Cell type: Microglia\"\n",
      "[1] \"Number of cells used: 1768\"\n",
      "[1] \"Gene-distance correlations:\"\n",
      "    gene correlation abs_correlation\n",
      "1  Trem2  -0.2954035       0.2954035\n",
      "2    Grn  -0.2932961       0.2932961\n",
      "3 Picalm  -0.2794015       0.2794015\n",
      "4   Cfl1   0.1074841       0.1074841\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(scater)\n",
    "library(ggpubr)\n",
    "library(pdist)\n",
    "library(viridis)\n",
    "library(dplyr)\n",
    "library(corrplot)\n",
    "library(ggcorrplot)\n",
    "\n",
    "# Data preprocessing\n",
    "expr_use_scater_norm <- normalizeCounts(expr_use)\n",
    "plaque <- read.csv(\"/media/data1/ai4bread_lab/lyfu/experiment/Celina/Data/plaque_13months-disease-replicate_1.csv\")\n",
    "\n",
    "# Gene expression heatmap based on cell type\n",
    "plot_gene_expression_heatmap_celltype <- function(gene_name, \n",
    "                                                 celltype_name = NULL,\n",
    "                                                 celltype_threshold = 0.5,\n",
    "                                                 grid_size = 80, \n",
    "                                                 smooth_radius = 40,\n",
    "                                                 output_file = NULL,\n",
    "                                                 width = 10, height = 8) {\n",
    "  \n",
    "  message(\"Creating expression heatmap for\", gene_name, \"in\", ifelse(is.null(celltype_name), \"all cells\", celltype_name), \"...\")\n",
    "  \n",
    "  # Check if gene exists\n",
    "  if (!gene_name %in% rownames(expr_use_scater_norm)) {\n",
    "    stop(paste(\"Gene\", gene_name, \"does not exist in expression matrix\"))\n",
    "  }\n",
    "  \n",
    "  # Filter cell types\n",
    "  if (!is.null(celltype_name)) {\n",
    "    if (!celltype_name %in% colnames(data_use$celltype_proportion)) {\n",
    "      available_types <- colnames(data_use$celltype_proportion)\n",
    "      stop(paste(\"Cell type\", celltype_name, \"does not exist. Available types:\", paste(available_types, collapse = \", \")))\n",
    "    }\n",
    "    \n",
    "    # Filter target cell types\n",
    "    celltype_prop <- data_use$celltype_proportion[, celltype_name]\n",
    "    target_cells <- which(celltype_prop >= celltype_threshold)\n",
    "    \n",
    "    if (length(target_cells) == 0) {\n",
    "      stop(paste(\"No cells meet the threshold\", celltype_threshold, \"for\", celltype_name))\n",
    "    }\n",
    "    \n",
    "    message(paste(\"Using\", length(target_cells), celltype_name, \"cells (threshold:\", celltype_threshold, \")\"))\n",
    "    \n",
    "    # Use only target cell type data\n",
    "    location_data <- data_use$location_use[target_cells, ]\n",
    "    gene_expr <- expr_use_scater_norm[gene_name, target_cells]\n",
    "  } else {\n",
    "    # Use all cells\n",
    "    location_data <- data_use$location_use\n",
    "    gene_expr <- expr_use_scater_norm[gene_name, ]\n",
    "    message(\"Using all\", nrow(location_data), \"cells\")\n",
    "  }\n",
    "  \n",
    "  # Create grid\n",
    "  x_range <- range(location_data[,1])\n",
    "  y_range <- range(location_data[,2])\n",
    "  \n",
    "  x_seq <- seq(x_range[1], x_range[2], length.out = grid_size)\n",
    "  y_seq <- seq(y_range[1], y_range[2], length.out = grid_size)\n",
    "  \n",
    "  grid_data <- expand.grid(x = x_seq, y = y_seq)\n",
    "  grid_data$expression <- NA\n",
    "  \n",
    "  # Calculate weighted average expression for each grid point\n",
    "  message(\"Calculating grid expression values...\")\n",
    "  for(i in 1:nrow(grid_data)) {\n",
    "    distances <- sqrt((location_data[,1] - grid_data$x[i])^2 + \n",
    "                     (location_data[,2] - grid_data$y[i])^2)\n",
    "    weights <- exp(-distances^2 / (2 * smooth_radius^2))\n",
    "    weights <- weights / sum(weights)\n",
    "    grid_data$expression[i] <- sum(gene_expr * weights, na.rm = TRUE)\n",
    "  }\n",
    "  \n",
    "  # Read plaque data\n",
    "  if (exists(\"plaque\")) {\n",
    "    plaque_data <- plaque\n",
    "  } else {\n",
    "    stop(\"Please ensure plaque data is loaded\")\n",
    "  }\n",
    "  \n",
    "  # Create plaque circles\n",
    "  create_plaque_circles <- function(plaque_data, n_points = 100) {\n",
    "    circle_data <- data.frame()\n",
    "    for(i in 1:nrow(plaque_data)) {\n",
    "      theta <- seq(0, 2*pi, length.out = n_points)\n",
    "      x_circle <- plaque_data$m.cx[i] + plaque_data$s.radius.mean[i] * cos(theta)\n",
    "      y_circle <- plaque_data$m.cy[i] + plaque_data$s.radius.mean[i] * sin(theta)\n",
    "      \n",
    "      temp_circle <- data.frame(\n",
    "        x = x_circle, y = y_circle, plaque_id = i,\n",
    "        radius = plaque_data$s.radius.mean[i]\n",
    "      )\n",
    "      circle_data <- rbind(circle_data, temp_circle)\n",
    "    }\n",
    "    return(circle_data)\n",
    "  }\n",
    "  \n",
    "  plaque_circles <- create_plaque_circles(plaque_data)\n",
    "  \n",
    "  # Create cell scatter layer\n",
    "  cell_layer <- NULL\n",
    "  if (!is.null(celltype_name)) {\n",
    "    cell_data <- data.frame(\n",
    "      x = location_data[,1],\n",
    "      y = location_data[,2],\n",
    "      expression = gene_expr\n",
    "    )\n",
    "    \n",
    "    cell_layer <- geom_point(data = cell_data, \n",
    "                            aes(x = x, y = y, color = expression), \n",
    "                            alpha = 0.6, size = 0.8)\n",
    "  }\n",
    "  \n",
    "  # Set title\n",
    "  title_text <- if (!is.null(celltype_name)) {\n",
    "    paste0(gene_name, \" in \", celltype_name, \" Cells\")\n",
    "  } else {\n",
    "    paste0(gene_name, \" Expression Heatmap\")\n",
    "  }\n",
    "  \n",
    "  # Plot heatmap\n",
    "  p <- ggplot() +\n",
    "    # Gene expression heatmap\n",
    "    geom_raster(data = grid_data, aes(x = x, y = y, fill = expression), \n",
    "                alpha = 0.75, interpolate = TRUE) +\n",
    "    scale_fill_viridis_c(option = \"D\", name = paste(gene_name, \"\\nExpression\")) +\n",
    "    \n",
    "    # Add cell scatter\n",
    "    cell_layer +\n",
    "    {if (!is.null(celltype_name)) scale_color_viridis_c(option = \"D\", guide = \"none\")} +\n",
    "    \n",
    "    # Plaque boundaries\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = \"white\", size = 3, alpha = 0.9) +\n",
    "    geom_path(data = plaque_circles, \n",
    "              aes(x = x, y = y, group = plaque_id), \n",
    "              color = \"red\", size = 2, alpha = 0.8) +\n",
    "    \n",
    "    # Theme settings\n",
    "    theme_void() +\n",
    "    theme(\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      text = element_text(size = 26),\n",
    "      legend.position = \"right\",\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.title = element_text(size = 26),\n",
    "      panel.background = element_rect(fill = \"white\")\n",
    "    ) +\n",
    "    ggtitle(title_text) +\n",
    "    coord_fixed()\n",
    "  \n",
    "  # Save\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, p, width = width, height = height, device = \"pdf\")\n",
    "    message(\"Heatmap saved to:\", output_file)\n",
    "  }\n",
    "  \n",
    "  return(list(\n",
    "    plot = p, \n",
    "    grid_data = grid_data,\n",
    "    cell_count = nrow(location_data)\n",
    "  ))\n",
    "}\n",
    "\n",
    "# Gene expression and distance correlation analysis based on cell type\n",
    "analyze_expression_distance_correlation_celltype <- function(gene_list, \n",
    "                                                           celltype_name = NULL,\n",
    "                                                           celltype_threshold = 0.5,\n",
    "                                                           output_file = NULL,\n",
    "                                                           method = \"spearman\") {\n",
    "  \n",
    "  celltype_text <- ifelse(is.null(celltype_name), \"all cells\", celltype_name)\n",
    "  message(\"Analyzing gene expression and plaque distance correlation in\", celltype_text, \"...\")\n",
    "  \n",
    "  # Read plaque data\n",
    "  if (exists(\"plaque\")) {\n",
    "    plaque_data <- plaque\n",
    "  } else {\n",
    "    stop(\"Please ensure plaque data is loaded\")\n",
    "  }\n",
    "  \n",
    "  # Filter cell types\n",
    "  if (!is.null(celltype_name)) {\n",
    "    if (!celltype_name %in% colnames(data_use$celltype_proportion)) {\n",
    "      available_types <- colnames(data_use$celltype_proportion)\n",
    "      stop(paste(\"Cell type\", celltype_name, \"does not exist. Available types:\", paste(available_types, collapse = \", \")))\n",
    "    }\n",
    "    \n",
    "    celltype_prop <- data_use$celltype_proportion[, celltype_name]\n",
    "    target_cells <- which(celltype_prop >= celltype_threshold)\n",
    "    \n",
    "    if (length(target_cells) == 0) {\n",
    "      stop(paste(\"No cells meet the threshold\", celltype_threshold, \"for\", celltype_name))\n",
    "    }\n",
    "    \n",
    "    message(paste(\"Using\", length(target_cells), celltype_name, \"cells\"))\n",
    "    \n",
    "    location_data <- data_use$location_use[target_cells, ]\n",
    "    expr_data <- expr_use_scater_norm[, target_cells]\n",
    "  } else {\n",
    "    location_data <- data_use$location_use\n",
    "    expr_data <- expr_use_scater_norm\n",
    "    message(\"Using all\", nrow(location_data), \"cells\")\n",
    "  }\n",
    "  \n",
    "  # Calculate distance from each cell to nearest plaque\n",
    "  calculate_distances <- function(location_data, plaque_data) {\n",
    "    distances <- apply(location_data, 1, function(cell_pos) {\n",
    "      min(apply(plaque_data[,1:2], 1, function(plaque_pos) {\n",
    "        sqrt(sum((cell_pos - plaque_pos)^2))\n",
    "      }))\n",
    "    })\n",
    "    return(distances)\n",
    "  }\n",
    "  \n",
    "  distances <- calculate_distances(location_data, plaque_data)\n",
    "  \n",
    "  # Check if genes exist\n",
    "  existing_genes <- gene_list[gene_list %in% rownames(expr_data)]\n",
    "  if (length(existing_genes) == 0) {\n",
    "    stop(\"No valid genes for analysis\")\n",
    "  }\n",
    "  \n",
    "  if (length(existing_genes) < length(gene_list)) {\n",
    "    missing_genes <- gene_list[!gene_list %in% rownames(expr_data)]\n",
    "    warning(\"The following genes do not exist:\", paste(missing_genes, collapse = \", \"))\n",
    "  }\n",
    "  \n",
    "  # Create gene expression data frame\n",
    "  gene_expr_data <- data.frame(row.names = 1:length(distances))\n",
    "  for (gene in existing_genes) {\n",
    "    gene_expr_data[[gene]] <- expr_data[gene, ]\n",
    "  }\n",
    "  \n",
    "  # Calculate inter-gene correlation matrix\n",
    "  gene_cor_matrix <- cor(gene_expr_data, method = method, use = \"complete.obs\")\n",
    "  \n",
    "  # Calculate gene-distance correlations\n",
    "  distance_correlations <- sapply(existing_genes, function(gene) {\n",
    "    cor(distances, expr_data[gene, ], method = method, use = \"complete.obs\")\n",
    "  })\n",
    "  \n",
    "  # Create gene-distance correlation results data frame\n",
    "  distance_cor_results <- data.frame(\n",
    "    gene = names(distance_correlations),\n",
    "    correlation = as.numeric(distance_correlations),\n",
    "    abs_correlation = abs(as.numeric(distance_correlations))\n",
    "  ) %>%\n",
    "    arrange(desc(abs_correlation))\n",
    "  \n",
    "  message(\"=== Gene-Plaque Distance Correlation Results in\", celltype_text, \"===\")\n",
    "  print(distance_cor_results)\n",
    "  \n",
    "  # Inter-gene correlation heatmap\n",
    "  p1 <- ggcorrplot(gene_cor_matrix, \n",
    "                   method = \"circle\",\n",
    "                   type = \"full\",\n",
    "                   ggtheme = theme_minimal(),\n",
    "                   title = paste(\"Gene-Gene Correlation in\", celltype_text, \"(\", method, \")\"),\n",
    "                   lab = FALSE, \n",
    "                   lab_size = 8) +\n",
    "    theme(\n",
    "      text = element_text( size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.title = element_text(size = 26),\n",
    "      axis.text.x = element_text(angle = 45, hjust = 1, size = 26),\n",
    "      axis.text.y = element_text(size = 26)\n",
    "    )\n",
    "  \n",
    "  # Gene-distance correlation dot plot\n",
    "  p2 <- ggplot(distance_cor_results, aes(x = reorder(gene, correlation), y = correlation)) +\n",
    "    geom_col(aes(fill = correlation > 0), alpha = 0.8, width = 0.7) +\n",
    "    geom_point(size = 3, color = \"black\") +\n",
    "    scale_fill_manual(values = c(\"TRUE\" = \"coral\", \"FALSE\" = \"skyblue\"),\n",
    "                      labels = c(\"Positive\", \"Negative\"), name = \"Correlation\") +\n",
    "    coord_flip() +\n",
    "    labs(x = \"Gene\", y = paste(\"Correlation with Distance (\", method, \")\"),\n",
    "         title = paste(\"Gene vs Distance in\", celltype_text)) +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      text = element_text( size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.title = element_text(size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      axis.text.x = element_text(size = 26),\n",
    "      axis.text.y = element_text(size = 26),\n",
    "      axis.title.x = element_text(size = 26),\n",
    "      axis.title.y = element_text(size = 26)\n",
    "    ) +\n",
    "    geom_hline(yintercept = 0, linetype = \"dashed\", alpha = 0.7, color = \"gray50\")\n",
    "  \n",
    "  # Scatter plots: Top 4 most correlated genes vs distance\n",
    "  top_genes <- head(distance_cor_results$gene, 4)\n",
    "  \n",
    "  scatter_plots <- list()\n",
    "  for (i in 1:length(top_genes)) {\n",
    "    gene <- top_genes[i]\n",
    "    correlation_val <- distance_cor_results$correlation[distance_cor_results$gene == gene]\n",
    "    \n",
    "    plot_data <- data.frame(\n",
    "      distance = distances,\n",
    "      expression = expr_data[gene, ]\n",
    "    )\n",
    "    \n",
    "    p_scatter <- ggplot(plot_data, aes(x = distance, y = expression)) +\n",
    "      geom_point(alpha = 0.6, size = 0.8, color = \"steelblue\") +\n",
    "      geom_smooth(method = \"lm\", color = \"red\", alpha = 0.8, se = TRUE) +\n",
    "      labs(x = \"Distance to Plaque\", y = paste(gene, \"Expression\"),\n",
    "           title = paste0(gene, \" vs Distance\\nr = \", round(correlation_val, 3))) +\n",
    "      theme_minimal() +\n",
    "      theme(\n",
    "        text = element_text( size = 26),\n",
    "        legend.text = element_text(size = 26),\n",
    "        legend.title = element_text(size = 26),\n",
    "        plot.title = element_text(size = 26, hjust = 0.5),\n",
    "        axis.text.x = element_text(size = 26),\n",
    "        axis.text.y = element_text(size = 26),\n",
    "        axis.title.x = element_text(size = 26),\n",
    "        axis.title.y = element_text(size = 26)\n",
    "      )\n",
    "    \n",
    "    scatter_plots[[i]] <- p_scatter\n",
    "  }\n",
    "  \n",
    "  p3 <- ggarrange(plotlist = scatter_plots, ncol = 2, nrow = 2)\n",
    "  \n",
    "  # Combine all plots\n",
    "  combined_plot <- ggarrange(\n",
    "    ggarrange(p1, p2, ncol = 2, widths = c(1.2, 1)),\n",
    "    p3,\n",
    "    nrow = 2, heights = c(1, 1)\n",
    "  )\n",
    "  \n",
    "  # Save\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, combined_plot, \n",
    "           width = 16, height = 12, device = \"pdf\")\n",
    "    message(\"Correlation analysis results saved to:\", output_file)\n",
    "  }\n",
    "  \n",
    "  return(list(\n",
    "    plot = combined_plot,\n",
    "    gene_correlation_matrix = gene_cor_matrix,\n",
    "    gene_distance_correlations = distance_cor_results,\n",
    "    gene_expression_data = gene_expr_data,\n",
    "    cell_count = nrow(location_data),\n",
    "    celltype_info = list(\n",
    "      celltype = celltype_name,\n",
    "      threshold = celltype_threshold,\n",
    "      cell_count = nrow(location_data)\n",
    "    )\n",
    "  ))\n",
    "}\n",
    "\n",
    "# Gene network correlation analysis based on cell type\n",
    "analyze_gene_network_celltype <- function(gene_list, \n",
    "                                         celltype_name = NULL,\n",
    "                                         celltype_threshold = 0.5,\n",
    "                                         correlation_threshold = 0.3,\n",
    "                                         output_file = NULL,\n",
    "                                         method = \"spearman\") {\n",
    "  \n",
    "  celltype_text <- ifelse(is.null(celltype_name), \"all cells\", celltype_name)\n",
    "  message(\"Analyzing gene co-expression network in\", celltype_text, \"...\")\n",
    "  \n",
    "  # Filter cell types\n",
    "  if (!is.null(celltype_name)) {\n",
    "    if (!celltype_name %in% colnames(data_use$celltype_proportion)) {\n",
    "      available_types <- colnames(data_use$celltype_proportion)\n",
    "      stop(paste(\"Cell type\", celltype_name, \"does not exist. Available types:\", paste(available_types, collapse = \", \")))\n",
    "    }\n",
    "    \n",
    "    celltype_prop <- data_use$celltype_proportion[, celltype_name]\n",
    "    target_cells <- which(celltype_prop >= celltype_threshold)\n",
    "    \n",
    "    if (length(target_cells) == 0) {\n",
    "      stop(paste(\"No cells meet the threshold\", celltype_threshold, \"for\", celltype_name))\n",
    "    }\n",
    "    \n",
    "    message(paste(\"Using\", length(target_cells), celltype_name, \"cells\"))\n",
    "    expr_data <- expr_use_scater_norm[, target_cells]\n",
    "  } else {\n",
    "    expr_data <- expr_use_scater_norm\n",
    "    message(\"Using all\", ncol(expr_data), \"cells\")\n",
    "  }\n",
    "  \n",
    "  # Check if genes exist\n",
    "  existing_genes <- gene_list[gene_list %in% rownames(expr_data)]\n",
    "  if (length(existing_genes) < 2) {\n",
    "    stop(\"At least 2 valid genes are required for correlation analysis\")\n",
    "  }\n",
    "  \n",
    "  # Create gene expression matrix\n",
    "  gene_expr_matrix <- t(expr_data[existing_genes, ])\n",
    "  colnames(gene_expr_matrix) <- existing_genes\n",
    "  \n",
    "  # Calculate correlation matrix\n",
    "  cor_matrix <- cor(gene_expr_matrix, method = method, use = \"complete.obs\")\n",
    "  \n",
    "  # Extract strongly correlated gene pairs\n",
    "  strong_correlations <- data.frame()\n",
    "  for(i in 1:(nrow(cor_matrix)-1)) {\n",
    "    for(j in (i+1):ncol(cor_matrix)) {\n",
    "      cor_val <- cor_matrix[i, j]\n",
    "      if(abs(cor_val) >= correlation_threshold) {\n",
    "        strong_correlations <- rbind(strong_correlations, data.frame(\n",
    "          Gene1 = rownames(cor_matrix)[i],\n",
    "          Gene2 = colnames(cor_matrix)[j],\n",
    "          Correlation = cor_val,\n",
    "          Abs_Correlation = abs(cor_val)\n",
    "        ))\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  if(nrow(strong_correlations) > 0) {\n",
    "    strong_correlations <- strong_correlations %>% arrange(desc(Abs_Correlation))\n",
    "    message(\"=== Strongly correlated gene pairs in\", celltype_text, \"(|r| >= \", correlation_threshold, \") ===\")\n",
    "    print(strong_correlations)\n",
    "  } else {\n",
    "    message(\"No strongly correlated gene pairs found in\", celltype_text, \"(threshold:\", correlation_threshold, \")\")\n",
    "  }\n",
    "  \n",
    "  # Correlation heatmap\n",
    "  p1 <- ggcorrplot(cor_matrix, \n",
    "                   method = \"circle\",\n",
    "                   type = \"upper\",\n",
    "                   ggtheme = theme_minimal(),\n",
    "                   title = paste(\"Gene Network in\", celltype_text, \"(\", method, \")\"),\n",
    "                   lab = FALSE,\n",
    "                   lab_size = 8,\n",
    "                   show.diag = TRUE) +\n",
    "    theme(\n",
    "      text = element_text( size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.title = element_text(size = 26),\n",
    "      axis.text.x = element_text(angle = 45, hjust = 1, size = 26),\n",
    "      axis.text.y = element_text(size = 26)\n",
    "    )\n",
    "  \n",
    "  # Correlation distribution histogram\n",
    "  cor_values <- cor_matrix[upper.tri(cor_matrix)]\n",
    "  p2 <- ggplot(data.frame(correlation = cor_values), aes(x = correlation)) +\n",
    "    geom_histogram(bins = 20, fill = \"skyblue\", color = \"black\", alpha = 0.7) +\n",
    "    geom_vline(xintercept = c(-correlation_threshold, correlation_threshold), \n",
    "               color = \"red\", linetype = \"dashed\", size = 1) +\n",
    "    labs(x = \"Correlation Coefficient\", y = \"Frequency\", \n",
    "         title = paste(\"Correlation Distribution in\", celltype_text)) +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      text = element_text( size = 26),\n",
    "      legend.text = element_text(size = 26),\n",
    "      legend.title = element_text(size = 26),\n",
    "      plot.title = element_text(size = 26, hjust = 0.5),\n",
    "      axis.text.x = element_text(size = 26),\n",
    "      axis.text.y = element_text(size = 26),\n",
    "      axis.title.x = element_text(size = 26),\n",
    "      axis.title.y = element_text(size = 26)\n",
    "    )\n",
    "  \n",
    "  # Strong correlation gene pairs bar plot\n",
    "  if(nrow(strong_correlations) > 0) {\n",
    "    strong_correlations$Pair <- paste(strong_correlations$Gene1, \"-\", strong_correlations$Gene2)\n",
    "    p3 <- ggplot(strong_correlations, aes(x = reorder(Pair, Abs_Correlation), \n",
    "                                         y = Correlation, fill = Correlation > 0)) +\n",
    "      geom_col(alpha = 0.8) +\n",
    "      scale_fill_manual(values = c(\"TRUE\" = \"coral\", \"FALSE\" = \"lightblue\"),\n",
    "                        labels = c(\"Positive\", \"Negative\"), name = \"Correlation\") +\n",
    "      coord_flip() +\n",
    "      labs(x = \"Gene Pairs\", y = \"Correlation Coefficient\",\n",
    "           title = paste0(\"Strong Correlations in \", celltype_text, \" (|r| ≥ \", correlation_threshold, \")\")) +\n",
    "      theme_minimal() +\n",
    "      theme(\n",
    "        text = element_text( size = 26),\n",
    "        legend.text = element_text(size = 26),\n",
    "        legend.title = element_text(size = 26),\n",
    "        plot.title = element_text(size = 26, hjust = 0.5),\n",
    "        axis.text.x = element_text(size = 23),\n",
    "        axis.text.y = element_text(size = 23),\n",
    "        axis.title.x = element_text(size = 26),\n",
    "        axis.title.y = element_text(size = 26)\n",
    "      ) +\n",
    "      geom_hline(yintercept = 0, linetype = \"dashed\", alpha = 0.5)\n",
    "  } else {\n",
    "    p3 <- ggplot() + \n",
    "      annotate(\"text\", x = 0.5, y = 0.5, \n",
    "               label = paste(\"No strong correlations found in\", celltype_text, \"\\n(threshold:\", correlation_threshold, \")\"),\n",
    "               size = 10) +\n",
    "      theme_void() +\n",
    "      labs(title = paste(\"Strong Correlations in\", celltype_text)) +\n",
    "      theme(plot.title = element_text(size = 26, hjust = 0.5))\n",
    "  }\n",
    "  \n",
    "  # Combine plots\n",
    "  combined_plot <- ggarrange(\n",
    "    p1,\n",
    "    ggarrange(p2, p3, ncol = 2, widths = c(1, 1)),\n",
    "    nrow = 2, heights = c(1.2, 1)\n",
    "  )\n",
    "  \n",
    "  # Save\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, combined_plot, \n",
    "           width = 14, height = 10, device = \"pdf\")\n",
    "    message(\"Gene network analysis results saved to:\", output_file)\n",
    "  }\n",
    "  \n",
    "  return(list(\n",
    "    plot = combined_plot,\n",
    "    correlation_matrix = cor_matrix,\n",
    "    strong_correlations = strong_correlations,\n",
    "    correlation_distribution = cor_values,\n",
    "    celltype_info = list(\n",
    "      celltype = celltype_name,\n",
    "      threshold = celltype_threshold,\n",
    "      cell_count = ncol(gene_expr_matrix)\n",
    "    )\n",
    "  ))\n",
    "}\n",
    "\n",
    "# Multi-gene heatmap comparison based on cell type\n",
    "plot_multi_gene_heatmaps_celltype <- function(gene_list, \n",
    "                                             celltype_name = NULL,\n",
    "                                             celltype_threshold = 0.5,\n",
    "                                             grid_size = 60,\n",
    "                                             smooth_radius = 35,\n",
    "                                             output_file = NULL,\n",
    "                                             ncol = 2) {\n",
    "  \n",
    "  celltype_text <- ifelse(is.null(celltype_name), \"all cells\", celltype_name)\n",
    "  message(\"Creating multi-gene heatmap comparison in\", celltype_text, \"...\")\n",
    "  \n",
    "  # Check genes\n",
    "  existing_genes <- gene_list[gene_list %in% rownames(expr_use_scater_norm)]\n",
    "  if (length(existing_genes) == 0) {\n",
    "    stop(\"No valid genes for analysis\")\n",
    "  }\n",
    "  \n",
    "  plot_list <- list()\n",
    "  \n",
    "  for (gene in existing_genes) {\n",
    "    result <- plot_gene_expression_heatmap_celltype(\n",
    "      gene_name = gene,\n",
    "      celltype_name = celltype_name,\n",
    "      celltype_threshold = celltype_threshold,\n",
    "      grid_size = grid_size,\n",
    "      smooth_radius = smooth_radius\n",
    "    )\n",
    "    plot_list[[gene]] <- result$plot\n",
    "  }\n",
    "  \n",
    "  # Combine plots\n",
    "  nrow <- ceiling(length(plot_list) / ncol)\n",
    "  combined_plot <- ggarrange(plotlist = plot_list, ncol = ncol, nrow = nrow)\n",
    "  \n",
    "  # Save\n",
    "  if (!is.null(output_file)) {\n",
    "    ggsave(output_file, combined_plot, \n",
    "           width = 6 * ncol, height = 5 * nrow, \n",
    "           device = \"pdf\")\n",
    "    message(\"Multi-gene heatmap saved to:\", output_file)\n",
    "  }\n",
    "  \n",
    "  return(combined_plot)\n",
    "}\n",
    "\n",
    "# Comprehensive analysis function based on cell type\n",
    "comprehensive_plaque_analysis_celltype <- function(gene_list,\n",
    "                                                  celltype_name = NULL,\n",
    "                                                  celltype_threshold = 0.5,\n",
    "                                                  correlation_threshold = 0.3,\n",
    "                                                  output_dir = \"comprehensive_analysis\") {\n",
    "  \n",
    "  celltype_text <- ifelse(is.null(celltype_name), \"all_cells\", celltype_name)\n",
    "  analysis_dir <- file.path(output_dir, paste0(\"analysis_\", celltype_text))\n",
    "  \n",
    "  if (!dir.exists(analysis_dir)) {\n",
    "    dir.create(analysis_dir, recursive = TRUE)\n",
    "  }\n",
    "  \n",
    "  message(\"=== Starting comprehensive plaque analysis for\", ifelse(is.null(celltype_name), \"all cells\", celltype_name), \"===\")\n",
    "  \n",
    "  # Generate heatmaps\n",
    "  message(\"1. Generating expression heatmaps...\")\n",
    "  heatmap_result <- plot_multi_gene_heatmaps_celltype(\n",
    "    gene_list = gene_list,\n",
    "    celltype_name = celltype_name,\n",
    "    celltype_threshold = celltype_threshold,\n",
    "    output_file = file.path(analysis_dir, \"expression_heatmaps.pdf\")\n",
    "  )\n",
    "  \n",
    "  # Correlation analysis\n",
    "  message(\"2. Performing gene-distance correlation analysis...\")\n",
    "  correlation_result <- analyze_expression_distance_correlation_celltype(\n",
    "    gene_list = gene_list,\n",
    "    celltype_name = celltype_name,\n",
    "    celltype_threshold = celltype_threshold,\n",
    "    output_file = file.path(analysis_dir, \"gene_distance_correlation.pdf\")\n",
    "  )\n",
    "  \n",
    "  # Gene network analysis\n",
    "  message(\"3. Performing gene co-expression network analysis...\")\n",
    "  network_result <- analyze_gene_network_celltype(\n",
    "    gene_list = gene_list,\n",
    "    celltype_name = celltype_name,\n",
    "    celltype_threshold = celltype_threshold,\n",
    "    correlation_threshold = correlation_threshold,\n",
    "    output_file = file.path(analysis_dir, \"gene_network_analysis.pdf\")\n",
    "  )\n",
    "  \n",
    "  # Save statistical results\n",
    "  write.csv(correlation_result$gene_distance_correlations, \n",
    "            file.path(analysis_dir, \"gene_distance_correlations.csv\"), \n",
    "            row.names = FALSE)\n",
    "  \n",
    "  write.csv(correlation_result$gene_correlation_matrix, \n",
    "            file.path(analysis_dir, \"gene_gene_correlation_matrix.csv\"), \n",
    "            row.names = TRUE)\n",
    "  \n",
    "  if(nrow(network_result$strong_correlations) > 0) {\n",
    "    write.csv(network_result$strong_correlations, \n",
    "              file.path(analysis_dir, \"strong_gene_correlations.csv\"), \n",
    "              row.names = FALSE)\n",
    "  }\n",
    "  \n",
    "  # Save analysis summary\n",
    "  analysis_summary <- data.frame(\n",
    "    Analysis_Type = c(\"Cell_Type\", \"Threshold\", \"Total_Cells\", \"Genes_Analyzed\", \"Strong_Correlations\"),\n",
    "    Value = c(\n",
    "      ifelse(is.null(celltype_name), \"All_Cells\", celltype_name),\n",
    "      celltype_threshold,\n",
    "      correlation_result$cell_count,\n",
    "      length(gene_list),\n",
    "      nrow(network_result$strong_correlations)\n",
    "    )\n",
    "  )\n",
    "  \n",
    "  write.csv(analysis_summary, \n",
    "            file.path(analysis_dir, \"analysis_summary.csv\"), \n",
    "            row.names = FALSE)\n",
    "  \n",
    "  message(\"=== Analysis completed ===\")\n",
    "  message(\"Results saved to:\", analysis_dir)\n",
    "  \n",
    "  return(list(\n",
    "    heatmaps = heatmap_result,\n",
    "    distance_correlations = correlation_result,\n",
    "    gene_network = network_result,\n",
    "    analysis_info = analysis_summary\n",
    "  ))\n",
    "}\n",
    "# Call functions\n",
    "gene_list <- c('Picalm', 'Cfl1', 'Grn', 'Trem2')\n",
    "cc <- \"Microglia\" \n",
    "\n",
    "# Single gene heatmap (optional individual analysis)\n",
    "heatmap_result <- plot_gene_expression_heatmap_celltype(\n",
    "  gene_name = \"Grn\",\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.3,\n",
    "  grid_size = 80,\n",
    "  smooth_radius = 40,\n",
    "  output_file = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/Grn_\", cc, \"_heatmap.pdf\")\n",
    ")\n",
    "\n",
    "# Gene-distance correlation analysis\n",
    "correlation_result <- analyze_expression_distance_correlation_celltype(\n",
    "  gene_list = gene_list,\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.1,\n",
    "  output_file = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/\", cc, \"_gene_distance_correlations.pdf\"),\n",
    "  method = \"spearman\"\n",
    ")\n",
    "\n",
    "\n",
    "# Multi-gene heatmap comparison (specific cell type)\n",
    "multi_heatmap <- plot_multi_gene_heatmaps_celltype(\n",
    "  gene_list = gene_list,\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.1,\n",
    "  output_file = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/\", cc, \"_multi_gene_heatmaps.pdf\"),\n",
    "  ncol = 3\n",
    ")\n",
    "\n",
    "# Comprehensive analysis\n",
    "comprehensive_result <- comprehensive_plaque_analysis_celltype(\n",
    "  gene_list = gene_list,\n",
    "  celltype_name = cc,\n",
    "  celltype_threshold = 0.1,\n",
    "  correlation_threshold = 0.3,\n",
    "  output_dir = paste0(\"/media/data1/ai4bread_lab/wkcui/experiments/Celina/Data/hip_results/\", cc, \"_comprehensive_analysis\")\n",
    ")\n",
    "\n",
    "# Display results\n",
    "print(\"=== Analysis completed ===\")\n",
    "print(paste(\"Cell type:\", cc))\n",
    "print(paste(\"Number of cells used:\", correlation_result$cell_count))\n",
    "print(\"Gene-distance correlations:\")\n",
    "print(correlation_result$gene_distance_correlations)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (CELINA)",
   "language": "R",
   "name": "r-celina"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
